\documentclass[a4paper, 11pt, twoside]{article}

% Basic packages
\usepackage{xltxtra} % Loads following 3 packages for processing with XeLaTeX
\usepackage{metalogo}
\usepackage{xunicode}
\usepackage{fontspec}

% Language selections
\usepackage{polyglossia}
% Colors
\usepackage[svgnames]{xcolor}
% In-line higlighting
\usepackage{soul}

% Modify page margins
\usepackage[a4paper, text={160mm, 250mm}]{geometry}

% Better captions of tables and figures
\usepackage[small, bf]{caption}

% More styles for lists
\usepackage{enumitem}

% More advanced tables
\usepackage{longtable} % For tables spanning multiple pages
\usepackage{array} % For vertical alignment within table cells

% Better bibliography
\usepackage[round, semicolon, authoryear]{natbib}
% \citet{} > Author (year); \citep{} > (Author, year); natbib: \citet/p[text][p.~1]{} > text Author year, p. 1; with * lists all authors
\bibliographystyle{plainnat}

% Various logos
\usepackage{hologo}

% Syntax higlight
\usepackage{minted}
% \usepackage{etoolbox}
\usemintedstyle{vim} % Styles are listed by pygmentize -L styles; languages are listed by pygmentize -L lexers
\newminted{bash}{bgcolor=Beige, firstnumber=last, fontfamily=tt, gobble=2, linenos, numbersep=-3pt} % Workarounds for older version of minted: \value{bashnumber}
\newmint{bash}{}
% Change line number style
\renewcommand{\theFancyVerbLine}{
  \sffamily
  \textcolor{BlueViolet}{
    \scriptsize
    \oldstylenums{
      \arabic{FancyVerbLine}
      }
    }
  }

% Add background for \texttt{} (using pacakge soul)
\sethlcolor{Beige}
\renewcommand{\texttt}[1]{\hl{\ttfamily #1}}

% Text in grey
\newcommand{\textgr}[1]{\textcolor{Grey}{#1}}

\setdefaultlanguage{english}

% Opening
\title{Manual for Sondovač 1.3} % TODO Update version
\author{Roswitha Schmickl, Aaron Liston, Vojtěch Zeisek and others}

% Allow line breaks within URLs
\usepackage[hyphens, lowtilde]{url}

% Clickable navigation within document
\usepackage[
  bookmarks=true,
  unicode=true,
  colorlinks=true,
  pagebackref=true,
  pdftitle={Sondovac manual 1.3 regular release}, % TODO Update version
  plainpages=false,
  pdfauthor={Roswitha Schmickl, Aaron Liston, Vojtech Zeisek and others},
  pdfsubject={Software manual},
  pdfcreator={XeTeX, http://xetex.sourceforge.net/},
  pdfkeywords={bioinformatics, genome skimming, manual, high-throughput sequencing},
  linkcolor={DarkRed},
  anchorcolor={DarkBlue},
  citecolor={DarkGreen},
  filecolor={DarkMagenta},
  menucolor={Brown},
  urlcolor={Blue},
  pdftex
  ]{hyperref}

\begin{document}

% Title page
\maketitle
\textbf{Sondovač}\footnote{English pronunciation is "Sondovach". The word is a~\href{https://en.wikipedia.org/wiki/Czech_language}{Czech} neologism meaning something like "The Prober" or "The Probe Maker".} is a~script to create orthologous low-copy nuclear probes from transcriptome and genome skim data for target enrichment \citep{Schmickl2016}. See \url{https://github.com/V-Z/sondovac/wiki}.

\begin{abstract}
Phylogenetics benefits from using a~large number of putatively independent nuclear loci in combination with other sources of information, such as the plastid and mitochondrial genome. Selecting such orthologous low-copy nuclear (LCN) loci is still a~challenge for non-model organisms. In recently published phylogenies based on target enrichment of several hundred LCN genes, these loci were selected from transcriptomes, genomes, gene expression studies, the literature, or a~combination of these sources. Automated bioinformatic pipelines for the selection of LCN genes are, however, largely absent. We created a~user-friendly, automated and interactive script named Sondovač to design LCN loci by a~comparison between transcriptome and genome skim data. The script is licensed under open-source license GPL v.3 allowing further modifications. It runs on major Linux distributions and Mac OS~X. Strong bioinformatics skills and access to high-performance computer clusters are not required; Sondovač runs on a~standard desktop computer equipped with modern CPU.
\end{abstract}

% Contents
\tableofcontents
\listoffigures
\listoftables
\vskip 1cm

% Document main text starts

\framebox{\parbox{15cm}{\textbf{Sondovač} is a~script to create orthologous low-copy nuclear probes from transcriptome and genome skim data for target enrichment \citep{Schmickl2016}. For information and download see \url{https://github.com/V-Z/sondovac/wiki}.}}
\vskip 0.75cm

\textbf{For newest version of this manual see \url{https://github.com/V-Z/sondovac/blob/master/manual/sondovac_manual.pdf}.}

\section{Introduction}

High-throughput sequencing (HTS) has the potential to greatly increase the amount of phylogenetically informative signal in molecular datasets \citep{Parks2009, Parks2012} and overcome difficulties in phylogenetic reconstructions, such as polytomies and low support values, that are often the result of using only a~small fraction of the genome. However, HTS also “opens the era of real incongruence” \citep{Jeffroy2006}, and even massive amounts of sequence data do not always result in strongly resolved phylogenies \citep{Pyron2015}.

Currently, target enrichment (sequence capture) of hundreds of loci is becoming increasingly popular in phylogenetics. In animal phylogenomics non-exonic or partly exonic ultraconserved elements and their more variable flanking regions are often utilized \citep[e.g.][]{Faircloth2012, Hedtke2013, Smith2014}. For plant phylogenetics, low-copy nuclear (LCN) genes are targeted \citep{Mandel2014, Weitemier2014, Grover2015, Heyduk2015, Mandel2015, Nicholls2015, Stephens2015a, Stephens2015} due to the paucity of ultraconserved nuclear sequences \citep{Reneker2012}. Target sequencing strategies for plant nuclear genomes are largely lineage-specific, requiring the de novo design of target enrichment probes. \citet{Chamala2015a} recently introduced a~pipeline for phylogenetic marker development in angiosperms using transcriptomes, and they obtained several hundred putative LCN genes that can be utilized at three phylogenetic levels (genus, family, order); however empirical evidence for the phylogenetic utility of these loci was not demonstrated. Alternative phylogenetic marker developments, also utilizing transcriptomes \citep{Pillon2014, Rothfels2013, Tonnabel2014}, resulted in a~much smaller number (up to 20) of mainly LCN loci, but these loci were evaluated with PCR in the empirical datasets, not target enrichment. In recently published phylogenies based on target enrichment of several hundred LCN genes, these loci were selected from transcriptomes, gene expression studies, the literature, or a~combination of these sources \citep{Mandel2014, Grover2015, Heyduk2015, Mandel2015, Nicholls2015, Stephens2015a, Stephens2015}. \citet{Weitemier2014} designed LCN probes for target enrichment based on a~combination of transcriptome and genome data. The limitation of this probe design pipeline is that (draft) genomes are still infrequent, especially for non-model species, and are costly to generate. This limitation also applies to the approach of \citet{DeSousa2014}, who selected 50 LCN loci from a~genomic source and amplified them using target enrichment. Except for \citet{Chamala2015a}, who offer a~user-friendly but empirically untested probe design pipeline, and \citet{Weitemier2014}, whose Hyb-Seq pipeline is designed for more advanced users, no automated probe design pipeline for LCN genes is currently available.

In this study \citep{Schmickl2016} we developed a~novel probe design pipeline for targeting orthologous LCN loci for phylogenetic reconstruction by using genome skim and transcriptome data. In particular, genome skim data of one accession of the studied plant group were combined with a~congeneric transcriptome from the 1000 Plants (1KP) initiative (\url{http://onekp.com/}). We implemented our software workflow in the user-friendly, automated and interactive BASH script Sondovač, which allows a~straightforward design of LCN probes also catering for users with limited bioinformatics skills.

Sondovač workflow is divided into three parts (see details on page~\pageref{pipeline-overview} and Figure~\ref{pipeline-workflow}):

\begin{enumerate}
\item Raw input data are analyzed by \texttt{sondovac$\_$part$\_$a.sh}.
\item Sequences obtained in part a~are assembled by Geneious in a~separate step by the user.
\item Final probes are produced by \texttt{sondovac$\_$part$\_$b.sh}. Eventually, not all plastid sequences will have been removed from the genome skim data by running \texttt{sondovac$\_$part$\_$a.sh}. In such case you will have to remove remaining plastid sequences manually (see detailed documentation of \texttt{sondovac$\_$part$\_$b.sh} on page~\pageref{partb}).
\end{enumerate}

\subsection{Pipeline -- how the data are processed}

A~transcriptome assembly and paired-end genome skim raw data are combined to get hundreds of orthologous LCN loci \citep{Schmickl2016}. Enrichment of multi-copy loci is minimized by using unique transcripts only, which are obtained by comparing all transcripts and removing those sharing $\geq$90\% sequence similarity using BLAT. Before matching the genome skim data against those unique transcripts, reads of plastid (and mitochondrial) origin are removed with Bowtie~2 and SAMtools, utilizing reference sequences. Paired-end reads are subsequently combined with FLASh. These processed reads are matched against the unique transcripts sharing $\geq$85\% sequence similarity with BLAT. Transcripts with $>$1000 BLAT hits (indicating repetitive elements) and BLAT hits containing masked nucleotides are removed before de novo assembly of the BLAT hits to larger contigs with Geneious, using the medium sensitivity / fast setting. After assembly, only those contigs that comprise exons of a~minimum bait length (usually $\geq$120 bp in case of probe design for phylogenies) and have a~certain minimum total locus length (multiple of the bait length, should not be too short in order to obtain sufficient phylogenetically informative signal; we recommend at least $\geq$600 bp) are retained. To ensure that probes do not target multiple similar loci, any probe sequences sharing $\geq$90\% sequence similarity are removed using cd-hit-est, followed by a~second filtering step for contigs containing exons of a~minimum bait length and totaling minimum loci length (see comments above). To ensure that plastid sequences are absent from the probes, the probe sequences are matched against the plastome reference sharing $\geq$90\% sequence similarity with BLAT and the hits removed from the probe set. The workflow of Sondovač is summarized in Figure~\ref{pipeline-workflow}. The direction of the workflow is indicated by arrows. An optional removal of reads of mitochondrial origin from the genome skim data is indicated by greyed text. The required input files of Sondovač are highlighted in bold.

\begin{figure}[p]
\begin{center}
\includegraphics[width=14.5cm]{pipeline_workflow.png}
\end{center}
\caption[Workflow of the probe design script Sondovač]{Workflow of the probe design script Sondovač. An overview of the main steps of Hyb-Seq are given in the top part of the figure; probe design is the first one. Each step of Sondovač is numbered and illustrated by three boxes: Software is highlighted in yellow, a~summary of each step is given in light blue, and input/output of each step is depicted in light green. An optional removal of reads of mitochondrial origin from the genome skim data is marked by greyed text. The required input files of Sondovač are highlighted in bold. The direction of the workflow is indicated by arrows. Eventually, not all plastid sequences will have been removed from the genome skim data in step 2 of the pipeline. In such case, you will have to remove remaining plastid sequences manually (see detailed documentation of \texttt{sondovac$\_$part$\_$b.sh} on page~\pageref{partb}).}
\label{pipeline-workflow}
\end{figure}

The steps of Sondovač are consecutively numbered to aid comprehension. Sondovač has three parts: two script parts and an intermediate part using Geneious. The workflow is as follows:

\begin{enumerate}[label=\textbf{\Alph*.}]
  \label{pipeline-overview}
  \item \texttt{sondovac$\_$part$\_$a.sh}: Covers steps 1~to 6.
    \begin{enumerate}[label=\textbf{\arabic*.}]
      \item Removal of transcripts sharing $\geq$90\% sequence similarity.
      \item Removal of reads of plastid origin.
      \item Removal of reads of mitochondrial origin (optional).
      \item Combination of paired-end reads.
      \item Matching of the unique transcripts and the filtered, combined genome skim reads sharing $\geq$85\% sequence similarity.
      \item Filtering of BLAT output:
	\begin{enumerate}[label=\textbf{6.\arabic*.}]
	\item Choice of transcript or genome skim sequences for further processing.
	\item Removal of transcripts with $>$1000 BLAT hits.
	\item Removal of transcript or genome skim BLAT hits [depending on the selection in (6.1)] containing masked nucleotides.
	\end{enumerate}
    \end{enumerate}

  Input files for \texttt{sondovac$\_$part$\_$a.sh} are FASTA transcriptome data, FASTQ paired-end genome skim reads and a~plastome (and possible also mitochondriome) reference. The input file for Geneious is the output of \texttt{sondovac$\_$part$\_$a.sh}.

  \item Geneious: Covers step 7~(see page~\pageref{geneious}).
    \begin{enumerate}[label=\textbf{\arabic*.}, resume]
    \item De novo assembly of the transcript or genome skim BLAT hits [depending on the selection in (6.1)] to larger contigs. Note that you need a~copy of Geneious for this step.
    \end{enumerate}

  The output files of Geneious are input files for \texttt{sondovac$\_$part$\_$b.sh}.

  \item \texttt{sondovac$\_$part$\_$b.sh}: Covers steps 8~to 11.
    \begin{enumerate}[label=\textbf{\arabic*.}, resume]
      \item Retention of those contigs that comprise exons $\geq$ bait length and have a~certain total locus length.
      \item Removal of probe sequences sharing $\geq$90\% sequence similarity.
      \item Retention of those contigs that comprise exons $\geq$ bait length and have a~certain total locus length.
      \item Detection of probe sequences sharing $\geq$90\% sequence similarity with the plastome reference. Eventually, not all plastid sequences will have been removed from the genome skim data by running \texttt{sondovac$\_$part$\_$a.sh}. In such case, you will have to remove remaining plastid sequences manually (see detailed documentation of \texttt{sondovac$\_$part$\_$\-b.sh} on page~\pageref{partb}).
    \end{enumerate}

  The output file of \texttt{sondovac$\_$part$\_$b.sh} is the final list of probes. In case of detection of remaining plastid sequences you will have to remove those plastid sequences manually from the output file of sondovac$\_$part$\_$b.sh (see detailed documentation of sondovac$\_$part$\_$b.sh on page~\pageref{partb}).

\end{enumerate}

When Sondovač starts, a~directory \texttt{bin} is created in the current working directory; Sondovač saves binaries of required software packages in this directory (if they are not available). The user can then add this directory to PATH, move or delete it afterwards.

\subsection{General considerations before you start}

The success of the probe design in terms of a high number of LCN genes of a sufficient minimum total length with Sondovač depends on various aspects of your transcriptome and genome skim input data:
\begin{itemize}
  \item number of transcripts,
  \item read length of genome skim reads; longer reads and paired-end reads are preferable due to a higher 	quality de novo assembly of the reads to contigs (exons),
  \item number of nuclear genome skim reads,
  \item quality of nuclear genome skim reads,
  \item sequence divergence between transcriptome and genome skim data.
\end{itemize}

These aspects influence the number of probe sequences and the proportion of paralogous loci among the probe sequences. The usage of a transcriptome and genome skim data of \textbf{diploid} accessions is strongly recommended in order to account for orthology of the probe sequences. An example of how one aspect, the number of nuclear genome skim reads, can affect the probe design, is shown in Table~\ref{summary-lcn-examples} and Figure~\ref{seq-div-examples}.

\begin{longtable}{ | >{\centering\arraybackslash}m{1.8cm} >{\centering\arraybackslash}m{6.5cm} >{\centering\arraybackslash}m{2.5cm} >{\centering\arraybackslash}m{3.4cm} |}
\caption[Summary of two examples of a LCN probe design with Sondovač.]{Summary of two examples of a LCN probe design with Sondovač. The \textit{Oxalis} example is from \citet{Schmickl2016}, the \textit{Curcuma} example is unpublished data from Tomáš Fér and Roswitha Schmickl. The respective Sondovač steps are listed; see Figure~\ref{pipeline-workflow} for details regarding these steps. For both probe designs 250~bp paired-end reads were utilized. Input files are given in \texttt{typewriter} font. Quality control of the genome skim data, which is not part of Sondovač, is colored in \textgr{grey}.}\\
\hline
\textbf{Step of Sondovač} & \textbf{Substep of Sondovač} & \textbf{\textit{Oxalis} species} & \textbf{\textit{Curcuma} species}\\
\endfirsthead % All the lines above this will be only on first page
\multicolumn{3}{@{}l}{\underline{\ldots~continued Table~\ref{summary-lcn-examples}.}}\\
\textbf{Step of Sondovač} & \textbf{Substep of Sondovač} & \textbf{\textit{Oxalis} species} & \textbf{\textit{Curcuma} species}\\
\endhead % All the lines above this will be repeated on every page
\hline
\endlastfoot
\texttt{Input file} & \texttt{Transcriptome taxon} & \textit{Oxalis corniculata} L. & \textit{Curcuma longa} L.\\
\texttt{Input file} & \texttt{Genome skim taxon} & \textit{Oxalis obtusa} Jacq. & \textit{Curcuma ecomata} Craib\\
\texttt{Input file} & \texttt{Plastome taxon} & \textit{Ricinus communis} L. & \textit{Curcuma roscoeana} Wall., \textit{Zingiber spectabile} Griff.\\
\texttt{Input file} & \texttt{Mitochondriome taxon} & \textit{Ricinus communis} L. & \textit{Oryza sativa} L. subsp. \textit{indica}\\
1 & Number of transcripts & 22,093 & 23,996\\
1 & Number of unique transcripts & 16,123 & 17,203\\
1 & Total length of unique transcripts & 11,799,393~bp & 11,919,459~bp\\
2 & Number of genome skim raw reads (without quality-filtering and duplicate removal) & 9,236,186 & 12,299,804\\
\textgr{Quality control} & \textgr{Percentage of dropped quality-filtered genome skim reads} & \textgr{2\%} & \textgr{2\%}\\
\textgr{Quality control} & \textgr{Number of quality-filtered genome skim reads} & \textgr{8,525,040} & \textgr{11,340,170}\\
\textgr{Quality control} & \textgr{Percentage of duplicate quality-filtered genome skim reads} & \textgr{7\%} & \textgr{3\%}\\
\textgr{Quality control} & \textgr{Number of quality-filtered genome skim reads after duplicate removal} & \textgr{7,938,349} & \textgr{11,041,405}\\
\textgr{Quality control} & \textgr{Number of masked bases in quality-filtered genome skim reads after duplicate removal} & \textgr{3,775} & \textgr{7,725}\\
3 & Number of nuclear genome skim raw reads (without quality-filtering and duplicate removal) & 8,240,470 & 11,636,852\\
4 & Number of combined nuclear genome skim raw reads & 2,619,197 & 3,834,278\\
4 & Combined nuclear genome skim raw reads as proportion of the total number of nuclear genome skim raw reads & 64\% & 66\%\\
4 & Total length of combined nuclear genome skim raw reads & 856,720,402~bp & 1,218,798,300~bp\\
5 & Mean sequence divergence between the unique transcripts and the combined nuclear genome raw skim reads & 7\% & 6\%\\
5 & Mean sequence length of the match between the unique transcripts and the combined nuclear genome raw skim reads (genome skim data) & 216~bp & 204~bp\\
5 & Mean sequence length of the match between the unique transcripts and the combined nuclear genome raw skim reads (transcripts) & 194~bp & 195~bp\\
7 & Mean sequence depth of the contigs (exons) after the de novo assembly of the matching sequences & 4 & 3\\
7 & Mean sequence length of the contigs (exons) after the de novo assembly of the matching sequences & 114~bp & 169~bp\\
7 & Mean pairwise identity between the assembled reads of the contigs (exons) after the de novo assembly of the matching sequences & 99\% & 100\%\\
7 & Minimum pairwise identity between the assembled reads of the contigs (exons) after the de novo assembly of the matching sequences & 84\% & 94\%\\
11 & Number of exons $\geq$120~bp & 4,926 & 4,618\\
11 & Number of genes & 1,164 ($\geq$600~bp) & 1,180 ($\geq$960~bp)\\
11 & Total length of probe sequences & 1,127,2049~bp & 1,571,800~bp
\label{summary-lcn-examples}
\end{longtable}

\begin{figure}[p]
\begin{center}
\includegraphics[width=15cm]{sequence_divergence_transcripts.png}
\end{center}
\caption[Sequence divergence examples.]{Sequence divergence between all transcripts and nuclear genome skim data (\textbf{A}-\textbf{D}) and unique transcripts and nuclear genome skim data (\textbf{E}, \textbf{F}) in the case of \textit{Oxalis} and \textit{Curcuma}. The generally larger number of \textit{Curcuma} nuclear genome skim reads compared to the \textit{Oxalis} nuclear genome skim reads is highlighted in red.}
\label{seq-div-examples}
\end{figure}

\section{Installation of Sondovač}
\label{install}

Sondovač is a~simple BASH script, but it requires additional software to run successfully. The script will check for the presence of all required software and, if needed, will offer installation. The easiest way is just to launch the script (see chapter \ref{script-start} on page \pageref{script-start}) and let yourself to be guided through the whole process.

\subsection{Requirements to run Sondovač}

Sondovač is currently tested on major Linux distributions (in current versions) openSUSE, Debian, Ubuntu, Linux Mint, Fedora, Centos and Scientific Linux; and on Mac OS~X (version 10.10 Yosemite).

In order to run Sondovač you need a~UNIX-based operating system (preferably Linux, alternatively Mac OS~X) equipped with BASH or a~compatible shell interpreter (this should by default be available for any Linux distribution, Mac OS~X and any other UNIX-based operating system like Solaris, BSD and its variants etc.). You should use the current operating system version supported by upstream, otherwise we will not be able to help you in case of problems. Older operating systems can have different versions of shell and system libraries, which can cause various problems and incompatibilities.

Sondovač uses several scientific software packages (namely BLAT, Bowtie2, CD-HIT, FLASh, Geneious, htsjdk, libgtextutils, and SAMtools -- see required versions and links, Table~\ref{software-links}), and basic UNIX tools (see below). Sondovač will check if those programs are installed -- available in the PATH (i.e. if the shell application can locate and launch respective binaries, see also vocabulary at page~\pageref{vocabulary}). If you have those packages installed (in current versions, see Table~\ref{software-links}), ensure that their binaries are in PATH. This should not be a~problem for basic tools available in any UNIX-based operating system, as basic installation usually contains all needed tools. If you lack some of the required tools, the script will notify you, and you will have to install them manually. If this is needed, check the documentation for your operating system.

If required programs are not installed, Sondovač will offer you installation. You can use precompiled binaries available together with the script (this is the recommended option) or (sometimes) from the web. In case you would like to compile required software yourself, the script will guide you through this process. This is recommended only for advanced users, as compilation might sometimes be very tricky. Users of Mac OS~X can install those applications also using Homebrew (see \url{https://brew.sh/}). For compilation you need, GNU G++, GNU GCC, GIT, libpng developmental files, and zlib developmental files. Ensure you have those tools available -- they should be readily available for any UNIX-based operating system. Chapters~\ref{required-linux} and~\ref{required-mac} give details about requirements and their manual fulfilling. This is mainly a~reference for more advanced users or users with special needs. For most users it should be fully sufficient to run the script and let it do this job (see chapter~\ref{script-start} on page~\pageref{script-usage}).

The following UNIX tools are required to run Sondovač. They are usually readily available in UNIX systems (but see note for Mac OS~X below), so there is usually no need to install them manually. The tools are \texttt{awk}, \texttt{bc}, \texttt{bunzip2}, \texttt{cat}, \texttt{cp}, \texttt{curl} or \texttt{wget}, \texttt{cut}, \texttt{dirname}, \texttt{dos2unix}, \texttt{echo}, \texttt{egrep}, \texttt{cd}, \texttt{g++}, \texttt{gcc}, \texttt{grep}, \texttt{gunzip}, \texttt{join}, \texttt{less}, \texttt{lsb$\_$release}, \texttt{make}, \texttt{mkdir}, \texttt{paste}, \texttt{perl}, \texttt{pwd}, \texttt{python}, \texttt{sed}, \texttt{sort}, \texttt{tar}, \texttt{tr}, \texttt{uname}, \texttt{uniq}, \texttt{unzip}, \texttt{wc}. Not all tools are required every time -- some are used only during particular actions (e.g. when the user decides to compile the required software manually). And the user usually does not need to bother with them. See also details in the following subchapters for some common Linux distributions and Mac OS~X.

See below for details about tools required by Sondovač and their manual installation. For most users it should be sufficient to be guided by the script to install needed tools automatically.

\subsection{Installation of required software in Linux}
\label{required-linux}

Linux distributions have precise package management tools (similar, but with more functions, to various app stores known from Android, iOS or recent Mac OS~X, MS~Windows, etc.), but unfortunately Linux repositories\footnote{On-line directories containing various software.} commonly do not contain all needed scientific packages (or not enough recent versions). We recommend to check if repositories of Linux distribution in use contain required scientific software and if not, use pre-compiled binaries of scientific applications available together with the script. If the user wishes to compile the software, for whatever reason, the script will guide through that process. Please note, that compilation may be complicated and require certain level of experience.

Following sections describe adding of extra repositories (if necessary) and installation of required scientific software on major Linux distributions. It is not part of the script itself, it must be done manually, and might require adjusting for particular Linux installation. For Linux users, the script offers usage of precompiled binaries or compilation of required software. Unfortunately, it is hard to cover variability of Linux distributions.

\subsubsection{openSUSE and SUSE Linux Enterprise (SLE)}

SUSE Linux Enterprise (\url{https://www.suse.com/}) and openSUSE (\url{https://www.opensuse.org/}) use for package management command \texttt{zypper}\footnote{See \url{https://en.opensuse.org/Zypper} and \url{https://activedoc.opensuse.org/book/opensuse-start-up/chapter-9-managing-software-with-command-line-tools} for details.}. The script will check if all required software packages are installed, and if not, will install them. You can also install manually:

\begin{bashcode}
  # Verify installation of basic tools (they are installed in 99.9%):
  sudo zypper in bash gawk bc coreutils grep less lsb-release perl-base python \
    sed wget
  # Install packages needed for compilation:
  sudo zypper in gcc-c++ gcc make bzip2 gzip tar unzip libpng12-devel \
    patterns-openSUSE-devel_basis zlib-devel dos2unix cpp
  # Update installed packages:
  sudo zypper up
  # Remove package:
  sudo zypper rm PACKAGE
  # Search for package:
  zypper se PACKAGE/KEYWORD
  # More information about zypper usage:
  zypper --help
  man zypper
  # Note backslash ("\") means that the code continues on the next line
\end{bashcode}

Originally, those distributions used only \texttt{rpm*} commands (see \texttt{rpm --help} and \texttt{man rpm} for basic usage).

For openSUSE, there is \href{https://en.opensuse.org/openSUSE:Science_Repositories}{Science Repository}. User can add and use it like this:

\begin{bashcode}
  # Help for adding new repository
  zypper ar -h
  # Add scientific repository containing Bowtie2 and SAMtools
  sudo zypper ar -r http://download.opensuse.org/repositories/science/` \
    lsb_release -d | cut -f 2 | sed 's/ /_/'`/science.repo -n science -e -f \
    -p 120
  # Install Bowtie2 and SAMtools (Blat, CD-HIT and FLASh are missing)
  sudo zypper in bowtie2 samtools
  # Note backslash ("\") means that the code continues on the next line
\end{bashcode}

\subsubsection{Debian, Ubuntu, Linux Mint and derivatives}

The biggest ``family'' of Linux distributions. Debian (\url{https://www.debian.org/}) (one of the odlest and biggest distributions), Linux Mint (\url{https://linuxmint.com/}), Ubuntu (\url{https://www.ubuntu.com/}) and all derived distributions\footnote{For complete lists see \url{https://distrowatch.com/search.php?basedon=Debian} and \url{https://distrowatch.com/search.php?basedon=Ubuntu}.} like Kubuntu (\url{https://kubuntu.com/}) use for package management commands \texttt{apt-get} (basic) and \texttt{aptitude} (text-based front-end for \texttt{apt-get}, recommended, not available by default in every DEB based distribution). There are more tools available\footnote{See \url{https://wiki.debian.org/PackageManagement} for list of tools and \url{https://www.debian.org/doc/manuals/debian-reference/ch02.en.html} for exhaustive documentation. A~shorter introduction is available at \url{https://help.ubuntu.com/community/AptGet/Howto} and \url{http://ubuntuguide.org/wiki/Ubuntu_Trusty_Packages_and_Repositories}. Ubuntu-specific information at \url{https://help.ubuntu.com/stable/ubuntu-help/addremove.html}.}, we will describe only the basic usage needed for our purpose. The script will check if all required software packages are installed, and if not, will install them. You can also install manually:

\begin{bashcode}
  # Verify installation of basic tools (they are installed in 99.9%):
  sudo apt-get install bash gawk bc coreutils grep less lsb-release perl-base \
    python sed wget
  # Install packages needed for compilation:
  sudo apt-get install build-essential bzip2 gzip tar unzip gcc g++ cpp make \
    libpng12-dev zlib1g-dev dos2unix
  # Update installed packages:
  sudo apt-get update # Update list of available packages in repositories
  sudo apt-get upgrade # Actually update installed packages
  # Remove package:
  sudo apt-get remove PACKAGE
  sudo apt-get autoremove # Automatically remove orphaned unneeded packages
  # Search for package:
  apt-cache --help # Usage options
  apt-cache show PACKAGE # Display information about PACKAGE
  apt-cache search KEYWORD # Search for KEYWORD, including regular expressions
  # More information about apt-get usage:
  apt-get --help
  man apt-get
  # Interactive command-line package manager
  sudo aptitude
  # Help for aptitude
  aptitude --help
  man aptitude
  # Note backslash ("\") means that the code continues on the next line
\end{bashcode}

Note you can use \texttt{aptitude} in a~similar way as \texttt{apt-*} commands (e.g. \texttt{aptitude instal PACKAGE} etc.). For special package operations, there are plenty of \texttt{dpkg} commands for advanced management.

Debian-based distributions have Bowtie2, CD-HIT and SAMtools (Blat and FLASh are missing) in their repositories. For Debian, it is readily installable, for Ubuntu it is necessary to enable \texttt{universe} repository by command \texttt{sudo add-apt-repository universe}. For graphical way and more details see \url{https://help.ubuntu.com/community/Repositories/Ubuntu}. Not not all Linux distributions derived from Debian and Ubuntu contain the packages. It is possible to add repositories from Debian or Ubuntu, but description is beyond this guide.

\begin{bashcode}
  # On Ubuntu and derivatives, allow universe repository
  sudo add-apt-repository universe
  # Install scientific packages
  sudo apt-get install bowtie2 cd-hit samtools
\end{bashcode}

\subsubsection{RedHat, Fedora, Centos, Scientific Linux and derivatives}

RedHat (\url{https://www.redhat.com/}), Fedora (\url{https://getfedora.org/}; until version 21), Centos (\url{https://centos.org/}) and Scientific Linux (\url{https://www.scientificlinux.org/}) and other related distributions\footnote{See \url{https://distrowatch.com/search.php?basedon=Fedora} for complete list.} use for package management command \texttt{yum}\footnote{See \url{http://yum.baseurl.org/} for details.}. The script will check if all required software packages are installed, and if not, will install them. You can also install manually:

\begin{bashcode}
  # Verify installation of basic tools (they are installed in 99.9%):
  sudo yum install bash coreutils gawk bc grep less lsb perl python sed wget
  # Install packages needed for compilation:
  sudo yum install bzip2 gzip unzip gcc gcc-c++ cpp libpng12-devel make \
    zlib-devel tar dos2unix
  # Update installed packages:
  sudo yum update
  # Remove package:
  sudo yum remove PACKAGE
  # Search for package:
  yum search PACKAGE/KEYWORD
  # More information about yum usage:
  yum --help
  man yum
  # Note backslash ("\") means that the code continues on the next line
\end{bashcode}

Since version 22, Fedora uses the command \texttt{dnf} for package management. It replaces older \texttt{yum}, and \texttt{yum} commands are redirected to \texttt{dnf}. The basic usage is the same, so that one can just replace \texttt{yum} with \texttt{dnf} in the above examples, see \url{https://dnf.readthedocs.io/en/latest/command_ref.html} for more usage of DNF on recent Fedora. Originally, those distributions used only \texttt{rpm*} commands (see \texttt{rpm --help} and \texttt{man rpm} for basic usage). Unfortunately, these disctributions do not contain much of required software (at least not in official repositories).

\subsection{Installation of required software in Mac OS~X}
\label{required-mac}

For Mac OS~X users, Homebrew (see \url{https://brew.sh/} and \url{https://github.com/Homebrew/}) will be installed by the script, and it will install (new software or newer versions) BASH (the shell interpreter), GNU AWK, GNU coreutils, GNU GCC, GNU grep, GNU make, GNU sed, and wget. Mac OS~X is missing some tools and contains outdated BSD versions for others (typically sed, grep or awk). The script will guide the user through the process, and the user can safely and easily remove these tools afterwards if necessary. Unfortunately, Mac OS~X does not have usable build-in package management, and it has outdated versions of some required tools. Homebrew fills this gap. It is a~simple command-line installer (similar to package managers known from Linux, BSD or Solaris) of various applications.

Homebrew requires Xcode\footnote{\url{https://developer.apple.com/xcode/}} (set of tools required to compile software) to be installed. Unfortunately, it is not possible to easily and universally check if Xcode is installed, so that the script will ask if the user wishes to install it. If the user is unsure if Xcode is installed, it is safe to answer \texttt{Yes} and install it. The manual command to install Xcode is the following:

\begin{bashcode}
  xcode-select --install
  # Following error means Xcode has already been installed:
  xcode-select: note: no developer tools were found at '/Applications/Xcode.app',
    requesting install. Choose an option in the dialog to download the command
    line developer tools.
  # Verify Xcode installation by
  xcode-select --print-path # Prints installation location of Xcode
  xcode-select --version # Prints version of Xcode
\end{bashcode}

If Xcode is not installed yet, the user will see windows similar to that on Figure~\ref{xcode}, offering installation of Xcode. Select \textbf{Install} to continue. After installation, the script will exit, and the user must start it again.

\begin{figure}[htb]
  \begin{center}
    \includegraphics[width=7.5cm]{xcode.png}
  \end{center}
  \caption[Prompt to install Xcode]{If the user uses the command requiring Xcode for the first time, the system will offer installation of Xcode.}
  \label{xcode}
\end{figure}

The script will guide the user through all those steps and basic usage of Homebrew. Manual installation of Homebrew is also simple:

\begin{bashcode}
  # Install Homebrew
  ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/ \
    master/install)"
  # Basic help
  brew help
  # Install UNIX tools required by Sondovač
  brew install coreutils gnu-sed gawk grep bash gcc make wget dos2unix python \
    perl
  # List of installed packages (brew formulae)
  brew list
  # Information about particular formula
  brew info FORMULA
  # Search for applications
  brew search KEYWORD
  # Update Homebrew
  brew update
  # Update all packages installed by Homebrew
  brew upgrade
  # Remove Homebrew package (formula)
  brew uninstall FORMULA
  # Cleaning after uninstallation
  brew cleanup
  # Completely remove Homebrew (after uninstallation of all formulae)
  ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/ \
    master/uninstall)"
  # Note backslash ("\") means that the code continues on the next line
\end{bashcode}

Within the Homebrew project, there is also a~scientific section (Homebrew Science, see \url{https://brew.sh/homebrew-science/}) containing plenty of software\footnote{See \url{https://github.com/Homebrew/homebrew-science/wiki/List-of-homebrew-science-formulae} for co\-mplete list of available scientific packages.}. When the script checks for required scientific packages, it offers several ways to install missing software. Mac OS~X users can also use Homebrew. It is the recommended way. If the user wishes to install the software manually, it is possible to use the following command:

\begin{bashcode}
  # Install required scientific packages using Homebrew Science
  brew install homebrew/science/blat homebrew/science/bowtie2 \
    homebrew/science/samtools homebrew/science/flash \
    homebrew/science/cd-hit
  # Note backslash ("\") means that the code continues on the next line
\end{bashcode}

Using Homebrew, software will be installed system-wide, and Homebrew easily allows checks for updates. Homebrew Science contains all software required by Sondovač.

\subsection{First launch of Sondovač}
\label{script-start}

Download the latest version from \url{https://github.com/V-Z/sondovac/releases/} and unpack the archive. You can run Sondovač from any directory. In command line navigate to the directory with the unpacked Sondovač files (see Figure~\ref{terminal}):

\begin{bashcode}
  cd /path/to/directory_with_sondovac
\end{bashcode}

and start it by

\begin{bashcode}
  ./sondovac_part_a.sh -h
\end{bashcode}

to see basic usage instructions. See chapter~\ref{script-usage} on page~\pageref{script-usage} for more information.

\begin{figure}[htbp]
  \includegraphics[width=\textwidth]{terminal.png}
  \caption[Starting terminal and navigating to Sondovač]{Starting terminal and navigating to Sondovač. First look at the terminal (command-line, shell) window, navigate to directory with Sondovač (using command \texttt{cd}), listing directory content (command \texttt{ls}) and preparing to launch Sondovač (\texttt{./sondovac$\_$part$\_$a.sh}).}
  \label{terminal}
\end{figure}

\subsubsection{Examples}

See page~\pageref{script-usage} for explanation of command line parameters. The basic and most simple usage (running in interactive mode, see chapter~\ref{script-usage} on page~\pageref{script-usage}):

\begin{bashcode}
  ./sondovac_part_a.sh -i
\end{bashcode}

Specify some of the required input files, otherwise run interactively:

\begin{bashcode}
  ./sondovac_part_a.sh -i -f input.fa -t reads1.fastq -q reads2.fastq
\end{bashcode}

Running in non-interactive, automated mode (parameter "\texttt{-n}", see chapter~\ref{script-usage} at page~\pageref{script-usage}) with 
example data downloaded from \url{https://github.com/V-Z/sondovac/wiki/Sample-data}:

\begin{bashcode}
  ./sondovac_part_a.sh -f input1_JHCN_Oxalis_corniculata_transcriptome_data.fa \
    -c input2_Ricinus_communis_reference_plastid_genome.fsa -m \
    input5_Ricinus_communis_reference_mitochondrial_genome.fasta -t \
    input3_J12_Oxalis_obtusa_genome_skim_data_R1.fastq -q \
    input4_J12_Oxalis_obtusa_genome_skim_data_R2.fastq -n
  # Note backslash ("\") means that the code continues on the next line
\end{bashcode}

Modify parameter "\texttt{-a}", otherwise run interactively:

\begin{bashcode}
  ./sondovac_part_a.sh -i -a 300
\end{bashcode}

Run in non-interactive mode (parameter "\texttt{-n}", see chapter~\ref{script-usage} at page~\pageref{script-usage}) -- in such cases the user must specify all required input files (parameters "\texttt{-f}", "\texttt{-c}", "\texttt{-m}", "\texttt{-t}" and "\texttt{-q}"). Moreover, parameter "\texttt{-y}" is modified:

\begin{bashcode}
  ./sondovac_part_a.sh -n -f input.fa -c referencecp.fasta \
    -m referencemt.fsa -t reads1.fastq -q reads2.fastq -y 90
\end{bashcode}

Modifying parameter "\texttt{-s}". Note that the interactive mode "\texttt{-i}" is implicit and does not need to be specified explicitly:

\begin{bashcode}
  ./sondovac_part_a.sh -s 950
\end{bashcode}

We recommend launching Sondovač in interactive mode,at least for the first time, so that the script can verify all requirements and install missing tools where needed. We recommend using non-interactive mode for routine usage.

\subsection{Help for usage of terminal}

If you are not familiar with the use of command line, try some basic tutorials first. Some options include:

\begin{itemize}
  \item \url{https://doc.opensuse.org/documentation/leap/startup/html/book.opensuse.startup/part.bash.html}
  \item \url{https://help.ubuntu.com/community/UsingTheTerminal}
  \item \url{https://www.gnu.org/software/bash/manual/} (advanced -- full reference manual)
  \item \url{https://www.debian.org/doc/manuals/debian-reference/ch01.en.html}
  \item \url{https://en.wikibooks.org/wiki/Guide_to_Unix}
  \item \url{http://tldp.org/LDP/Bash-Beginners-Guide/html/Bash-Beginners-Guide.html}
  \item \url{https://trapa.cz/en/course-linux-command-line-2016}
  \item \url{http://linuxcourse.rutgers.edu/documents/Bash-Beginners-Guide/}
  \item \url{http://ryanstutorials.net/linuxtutorial/}
  \item \url{http://www.hypexr.org/bash_tutorial.php}
  \item \url{http://mywiki.wooledge.org/BashGuide}
  \item \url{https://docs.fedoraproject.org/f26/system-administrators-guide/}
\end{itemize}

\subsection{Geneious}
\label{geneious}

For part \textbf{B} of the script the user must have Geneious \citep{Kearse2012}. Geneious is a~DNA alignment, assembly, and analysis software and one of the most common software platforms used in genomics. It is utilized for de novo assembly in Sondovač. We plan to replace it with a free open-source command line tool in a future release of Sondovač. Visit \url{https://www.geneious.com/} for download, purchase, installation and usage of Geneious. After the input data are processed (interactively or not) by \texttt{sondovac$\_$part$\_$a.sh}, the user must process its output manually with Geneious according to the instructions given below. The output of Geneious is then processed by \texttt{sondovac$\_$part$\_$b.sh}, which produces the final probe set. Geneious versions 6, 7~and 8 have been tested and are compatible with this script.

\subsection{Software used by Sondovač}

Table~\ref{software-links} lists all software used by Sondovač, including minimal required versions and homepages. As long as you have a~recently-updated version of your operating system and you use the automated installation of additional software offered by Sondovač, you do not need to worry about this. In case you installed some of the required scientific packages manually, ensure that you have the required minimal version. The following list refers to papers and web resources describing methods used by software utilized by Sondovač:

\begin{longtable}{| >{\centering\arraybackslash}m{2.8cm} >{\centering\arraybackslash}m{1.5cm} >{\centering\arraybackslash}m{10cm} |}
\caption[Required software, its versions and homepages.]{Required software, its versions and homepages. "X" denotes any subversion of particular lineage and "v. $>$" denotes any version higher than noted. Generally, any current version should usually be fine.}\\
\hline
\textbf{Software} & \textbf{Version} & \textbf{Homepage}\\
\endfirsthead % All the lines above this will be only on first page
\multicolumn{3}{@{}l}{\underline{\ldots~continued Table~\ref{software-links}.}}\\
\textbf{Software} & \textbf{Version} & \textbf{Homepage}\\
\endhead % All the lines above this will be repeated on every page
\hline
\endlastfoot
BASH & v. > 4 & \url{https://gnu.org/software/bash/bash.html}\\
BLAT & v.36 & \url{https://genome.ucsc.edu/FAQ/FAQblat.html}\\
Bowtie2 & 2.2.6 & \url{http://bowtie-bio.sourceforge.net/bowtie2/index.shtml}\\
CD-HIT & 4.6 & \url{http://weizhongli-lab.org/cd-hit/}\\
FLASh & 1.2.11 & \url{https://sourceforge.net/projects/flashpage/}\\
G++, GCC & v. > 4.2 & \url{https://gcc.gnu.org/}\\
Geneious & v. > 6.1 & \url{https://www.geneious.com/}\\
GNU core utils & 8.X & \url{https://gnu.org/software/coreutils/coreutils.html}\\
grab$\_$synglet\-on$\_$clusters.py & 1.00 & \url{https://github.com/listonlab/Hyb-Seq_protocol/}\\
libpng & 1.2.X & \url{http://www.libpng.org/}\\
SAMtools, htsjdk & 1.2 & \url{http://www.htslib.org/}\\
Sondovač & 1.3 & \url{https://github.com/V-Z/sondovac/wiki}\\
zlib & 1.2.8 & \url{https://zlib.net/}
\label{software-links}
\end{longtable}

\texttt{sondovac$\_$part$\_$a.sh} requires (and will install) the following software packages:

\begin{itemize}
  \item BLAT
  \item Bowtie2
  \item SAMtools
  \item FLASh
\end{itemize}

\texttt{sondovac$\_$part$\_$b.sh} requires (and will install) the following software packages:

\begin{itemize}
  \item CD-HIT
  \item BLAT
  \item grab$\_$syngleton$\_$clusters.py (included with Sondovač)
\end{itemize}

Papers describing the software used by Sondovač:

\begin{description}
 \item[BLAT] \citet{Kent2002}: BLAT -- the BLAST-like alignment tool.
 \item[Bowtie2] \citet{Langmead2012}: Fast gapped-read alignment with Bowtie 2.
 \item[CD-HIT] There are several papers describing CD-HIT:
  \begin{itemize}
    \item \citet{Li2001}: Clustering of highly homologous sequences to reduce the size of large protein databases.
    \item \citet{Li2002}: Tolerating some redundancy significantly speeds up clustering of large protein databases.
    \item \citet{Li2006}: Cd-hit: a~fast program for clustering and comparing large sets of protein or nucleotide sequences.
    \item \citet{Fu2012}: CD-HIT: accelerated for clustering the next generation sequencing data.
    \item \citet{Huang2010}: CD-HIT Suite: a~web server for clustering and comparing biological sequences.
    \item \citet{Niu2010}: Artificial and natural duplicates in pyrosequencing reads of metagenomic data.
    \item \citet{Li2012b}: Ultrafast clustering algorithms for metagenomic sequence analysis.
   \end{itemize}
 \item[FLASh] \citet{Magoc2011}: FLASh: fast length adjustment of short reads to improve genome assemblies.
 \item[Geneious] \citet{Kearse2012}: Geneious Basic: An integrated and extendable desktop software platform for the organization and analysis of sequence data.
 \item[grab$\_$syngleton$\_$clusters.py] \citet{Weitemier2014}: Hyb-Seq: Combining target enrichment and genome skimming for plant phylogenomics.
 \item[SAMtools] There are several papers describing SAMtools:
  \begin{itemize}
   \item \citet{Li2009}: The Sequence alignment/map (SAM) format and SAMtools.
   \item \citet{Li2011}: A~statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data.
   \item \citet{Li2011a}: Improving SNP discovery by base alignment quality.
  \end{itemize}
 \item[Sondovač] \citet{Schmickl2016}: Phylogenetic marker development for target enrichment from transcriptome and genome skim data: the pipeline and its application in southern African \textit{Oxalis} (Oxalidaceae).
\end{description}

\subsection{The PATH variable}

PATH (\$PATH) is a~system variable used in every UNIX system. It lists directories (separated by colon ":") where the current shell (see also Chapter~\ref{vocabulary} Vocabulary at page~\pageref{vocabulary}) searches for binaries (commands), so that the user does not have to specify the full path to the software (e.g. just \texttt{sed} instead of \texttt{/usr/bin/sed}). If some software is installed outside standard locations, the user must specify the full path, even if the user is located in the same directory as the software (e.g. \texttt{./sondovac$\_$part$\_$a.sh} -- this is for security reasons). In the case of two commands with the same name (e.g. \texttt{/bin/somecommand} and \texttt{/usr/bin/somecommand}), the order of directories in \$PATH matters -- the first occurrence is used, and any later commands are ignored (but this is usually a~rare case). PATH can be managed using the following commands:

\begin{bashcode}
  # See the $PATH variable
  echo $PATH # Sample output is on the next line:
  /home/$USER/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/sbin:/usr/sbin
  # Adding new directory to $PATH
  export PATH=$PATH:/some/new/directory
  # Do not do it in the following way - it would overwrite $PATH, and
  #   there would be only the new directory (not the original content)!
  export PATH=/some/new/directory # Wrong! Old PATH is missing and will be lost!
  # Removing possible duplicate entries in PATH with regular expressions and awk
  export PATH="$(echo "$PATH" | awk 'BEGIN{RS=":";}{sub(sprintf("%c$",10),""); \
    if(A[$0]){}else{A[$0]=1;printf(((NR==1)?"":":")$0)}}')"
  echo $PATH # See it after modifications
  # Note backslash ("\") means that the code continues on the next line
\end{bashcode}

Sondovač requires certain software to be installed (see Table~\ref{software-links} on page~\pageref{software-links}), and if some software is missing, the script offers installation. By default, Sondovač creates a~directory \texttt{bin} in the current working directory in which it installs the required software. It then temporarily modifies the content of the \$PATH variable to contain this new directory. Sondovač notifies the user about this, and the user can then -- if it is wished to keep the newly installed software for later usage -- (1) move the content of this directory to a preferred location or (2) add this directory to the \$PATH. This directory can also be safely removed. Permanent modification of the \$PATH variable is done by adding line \texttt{export PATH=\$PATH:/some/new/directory} (same as the above example) to file \texttt{$\sim$/.bashrc} or \texttt{$\sim$/.bash$\_$profile} (usage of those files varies slightly among UNIX systems, see manual for your operating system). On Mac OS~X, installation of \href{https://brew.sh/}{Homebrew} is required. For correct functioning of Sondovač, \$PATH is modified to contain directories \texttt{/usr/local/opt/coreutils/libexec/gnubin} and \texttt{/usr/local/bin} containing new and updated UNIX tools. The first directory must always be present, as it contains updated versions of basic command line utilities -- replacing outdated versions provided with Mac OS~X. All those modifications are temporary and used only within Sondovač scripts.

\subsection{Vocabulary}
\label{vocabulary}

\begin{description}
  \item[Binary] An application in a~form that is understandable by the computer, but usually not transferable among operating systems and/or hardware platforms. Binaries in Windows usually have the extension *.exe, in UNIX there is usually no extension.
  \item[BASH] "The command line" -- fully featured programming scripting language accessible through the terminal of any UNIX-based operating system (any Linux, Mac OS~X, Solaris, any variant of BSD and more). BASH scripts usually have the extension *.sh.
  \item[BSD] Group of popular UNIX-based operating systems. See \url{https://en.wikipedia.org/wiki/Berkeley_Software_Distribution}.
  \item[C] Popular programming language. Source code must be compiled for each operating system. See \url{https://en.wikipedia.org/wiki/C_(programming_language)}.
  \item[C++] Popular programming language. Source code must be compiled for each operating system. See \url{https://en.wikipedia.org/wiki/C++}.
  \item[Centos] Popular Linux distribution. Community remake of RedHat Enterprise Linux. See \url{https://centos.org/}.
  \item[Compilation] "Translation" of software application from the source code (text readable by human programmer) into binary form launchable by the computer. It requires special tools (compilers), and it usually must be done for every operating system and hardware platform.
  \item[Console] See "Shell".
  \item[Debian] One of the oldest and most popular Linux distributions. See \url{https://www.debian.org/}.
  \item[Fedora] Popular Linux distribution developed together with RedHat Linux as its free community testing platform. See \url{https://getfedora.org/}.
  \item[GNU] Major project providing free software widely used in many operating systems, see \url{https://gnu.org/}.
  \item[Homebrew] Tool primarily for Mac OS~X (although there is also a~Linux version available) replacing the virtually missing package manager for this system. Can be used to install plenty of various applications as well as updating tools already available in Mac OS~X. See \url{https://brew.sh/}.
  \item[Library] Pack of software tools and functions used by other applications.
  \item[License] Conditions under which software is distributed. Can be very restrictive (typically paid software) or permissive (typically free and open-source software).
  \item[Linux] One of the most common variants of UNIX-based operating systems. Linux kernel is used by many developers, so that there are plenty of Linux distributions ("flavors") from various sources (e.g. Ubuntu and derivatives, openSUSE, SLE, Debian, Linux Mint, Fedora, Centos, RedHat etc.). They share many features, although at first sight they can look different. See \url{https://en.wikipedia.org/wiki/Linux}.
  \item[Linux Mint] Popular Linux distribution based on Ubuntu (mainly) and Debian, see \url{https://linuxmint.com/}.
  \item[Mac OS~X] Popular operating system produced by Apple. The system kernel is based on UNIX, see \url{https://www.apple.com/osx/}.
  \item[Open-source] Generally, the source code of an application is available together with the application and can, under certain conditions, be defined in license modified, redistributed etc. See \url{https://en.wikipedia.org/wiki/Free_and_open-source_software}.
  \item[openSUSE] Popular Linux distribution, see \url{https://www.opensuse.org/}.
  \item[Operating system] Basic system running on your computer -- typically MS Windows (not supported by Sondovač, although it might work), Mac OS~X or some Linux distribution (Ubuntu and derivatives, openSUSE, SLE, Debian, Linux Mint, Fedora, Centos, RedHat etc.).
  \item[Package] Software or its part, group of tools, library etc. Basic unit of software management in most UNIX systems (mainly Linux, Solaris, BSD, practically missing in Mac OS~X). Those systems usually have special applications (command line as well as graphical tools) to easily manage (install, remove, update) software.
  \item[Parameter(s)] Option(s) passed to any function/command line application to modify its usage. Some can be required, some are optional, and some can be used only in particular cases. In case of shell applications, parameters are usually given such as "application -X", "application -parameter", "application -Param SomeValue" and so on. See manual for particular application (e.g. "man application"), in case of Sondovač see page~\pageref{script-usage}.
  \item[PATH] Directories in the computer where the system looks for installed software (in a~UNIX-based system you can view it by the command "\texttt{echo \$PATH}"). If you need to modify it manually, see the documentation for your operating system.
  \item[Perl] Popular interpreted programming language excelling mainly in system tasks working with text. Perl scripts are easily transferable among operating systems. See \url{https://www.perl.org/}.
  \item[Python] Very popular and powerful interpreted programming language used for wide variety of tasks. Python scripts are easily transferable among operating systems. See \url{https://www.python.org/}.
  \item[RedHat] Probably the biggest Linux company providing mainly solutions for big companies. See \url{https://www.redhat.com/}.
  \item[Repository] Internet folder (available through HTTP or FTP) containing software packages for UNIX systems.
  \item[Scientific Linux] Popular Linux distribution. Community remake of RedHat Enterprise Linux. See \url{https://www.scientificlinux.org/}.
  \item[Script] Software application. It requires an interpreter (application installed on the computer that is able to launch scripts written in a~particular language), but the application itself is portable among operating systems and hardware architectures, and it is written in plain text, so that developers can easily modify it. Common examples are Python, Perl or BASH.
  \item[Shell] "The command line" -- the interface to interact with software using commands typed into the terminal window (See Figure~\ref{terminal}).
  \item[Solaris] Popular (mainly on servers) UNIX-based operating system, now developed by Oracle and including several independent clones. See \url{https://distrowatch.com/table.php?distribution=solaris}.
  \item[Source code] Human-readable code written in any text editor used to develop any application. Applications written in interpreted languages (BASH, Perl, Python , \ldots) can be distributed just in form of a~source code (nothing else is required). Other programming languages (C, C++, \ldots) require compilation to get a~fully functional application.
  \item[SUSE Linux Enterprise (SLE)] Large Linux company providing mainly solutions for big companies. See \url{https://www.suse.com/}.
  \item[Terminal] see "Shell".
  \item[Ubuntu] Popular Linux distribution, see \url{https://www.ubuntu.com/}. There are plenty of distributions based on Ubuntu. See \url{https://distrowatch.com/search.php?basedon=Ubuntu}.
  \item[UNIX (UNIX-like, UN*X, *nix,~\ldots)] Family of operating systems sharing the same logic, software architecture and plenty of tools. See \url{https://en.wikipedia.org/wiki/Unix-like} for details.
  \item[Upstream] Developers usually support (e.g. by fixing of bugs) only newer versions of an application. If you use an older version and you encounter problems, no one can probably help you. Moreover, using old versions of software can be a~security risk because of security issues fixed in newer versions.
  \item[Variable] Named value storing various information, one of the basic part of any programming language, application, operating system.
\end{description}

\section{Usage of Sondovač}
\label{script-usage}

\subsection{Command line parameters}

Sondovač has some parameters that are useful especially for advanced users, on remote servers, for repeated analyses and so on. We recommend starting with basic interactive usage -- the script will ask for the input files and, if needed, also for installation of additional software.

\begin{bashcode}
  # Go to directory with unpacked Sondovač (in terminal):
  cd /path/to/directory_with_sondovac
  # Run sondovac_part_a.sh in basic interactive mode:
  ./sondovac_part_a.sh -i
  # Then run Geneious and continue with sondovac_part_b.sh:
  ./sondovac_part_b.sh -i
\end{bashcode}

\subsubsection{General parameters}

Shared by \texttt{sondovac$\_$part$\_$a.sh} as well as \texttt{sondovac$\_$part$\_$b.sh}.

\begin{description}
  \item[\texttt{-h}, \texttt{-v}] Print help message and exit.
  \item[\texttt{-u}] Check for updates. If there is a~newer version of Sondovač available on \url{https://github.com/V-Z/sondovac/releases/}, download of the newer version will be offered to the user.
  \item[\texttt{-l}] Display LICENSE for license information (this script is licensed under GNU GPL v.3, other software under variable licenses). Exit viewing by pressing the \texttt{Q} key.
  \item[\texttt{-r}] Display README for detailed usage instructions. Exit viewing by pressing the \texttt{Q} key.
  \item[\texttt{-p}] Display INSTALL for detailed installation instructions. Exit viewing by pressing the \texttt{Q} key. See also page~\pageref{install}.
  \item[\texttt{-e}] Display detailed citation information and exit.
  \item[\texttt{-o}] Set name of output files. Output files will start with that name. Do not use spaces or special characters - some software can't handle them correctly. The default value (if the user does not provide another name) is "output". See below for the list of produced output files.
  \item[\texttt{-i}] Running in interactive mode -- the script will on-demand ask for the required input files, installation of missing software etc.. This is the recommended default value (the script runs interactively without explicitly using option \texttt{-n}).
  \item[\texttt{-n}] Running in non-interactive mode. The user must provide at least the required input files (see below). You can use only one of the parameters \texttt{-i} or \texttt{-n} (not both of them). If the script fails to find some of the required software packages, it will exit. This is recommended for batch or repeated analysis, on remote servers and for more advanced users. The user must be sure that all required software is installed (see page~\pageref{install}).
\end{description}

\subsubsection{Input files}

Those parameters are required when running the script in non-interactive mode. The parameters are optional in default interactive mode. Please use file names without spaces and without special characters.

\begin{description}
\item[\texttt{-f FILE}] Transcriptome input file in FASTA format.
  \begin{itemize}
    \item \texttt{sondovac$\_$part$\_$a.sh}
  \end{itemize}
\item[\texttt{-c FILE}] Plastome reference sequence input file in FASTA format.
  \begin{itemize}
    \item \texttt{sondovac$\_$part$\_$a.sh}, \texttt{sondovac$\_$part$\_$b.sh}
    \item Plastome reference sequences from taxa up to the same order of the studied plant group are suitable. See \citet{Straub2012}.
  \end{itemize}
\item[\texttt{-m FILE}] Mitochondriome reference sequence input file in FASTA format (optional).
  \begin{itemize}
    \item \texttt{sondovac$\_$part$\_$a.sh}
    \item This step is optional, as plant mitochondrial genomes have largely variable sizes and high rearrangement rates.
  \end{itemize}
\item[\texttt{-t FILE}] Paired-end genome skim input file in FASTQ format (first file).
  \begin{itemize}
    \item \texttt{sondovac$\_$part$\_$a.sh}
  \end{itemize}
\item[\texttt{-q FILE}] Paired-end genome skim input file in FASTQ format (second file).
  \begin{itemize}
    \item \texttt{sondovac$\_$part$\_$a.sh}
  \end{itemize}
\item[\texttt{-x FILE}] Input file in TSV format (output of Geneious assembly).
  \begin{itemize}
    \item \texttt{sondovac$\_$part$\_$b.sh}
  \end{itemize}
\item[\texttt{-z FILE}] Input file in FASTA format (output of Geneious assembly).
  \begin{itemize}
    \item \texttt{sondovac$\_$part$\_$b.sh}
  \end{itemize}
\end{description}

\subsubsection{Optional parameters}

See page~\pageref{pipeline-overview} and Figure~\ref{pipeline-workflow} for steps referred here. If those parameters are not provided, the default values are used, and it is not possible to change them any time later (not even in interactive mode).

\begin{description}
\item[\texttt{-a \#\#\#}] Maximum overlap length expected in approximately $\geq$90\% of read pairs (parameter -M of FLASh, see its manual for details).
  \begin{itemize}
    \item Step 4~of Sondovač, \texttt{sondovac$\_$part$\_$a.sh}.
    \item FLASh can not combine paired-end reads that do not overlap by at least 10 bp (default minimum overlap length).
    \item DEFAULT: 65
    \item OPTIONS: Integer ranging from 10 to 300
  \end{itemize}
\item[\texttt{-y \#\#}] Sequence similarity between unique transcripts and the filtered, combined genome skim reads (parameter -minIdentity of BLAT, see its manual for details).
  \begin{itemize}
    \item Step 5~of Sondovač, \texttt{sondovac$\_$part$\_$a.sh}.
    \item Filtering for orthologs, using sequence similarity as criterion.
    \item DEFAULT: 85 (highly recommended)
    \item OPTIONS: Integer ranging from 70 to 100
  \end{itemize}
\item[\texttt{-g}] Choice of transcript or genome skim sequences for further processing.
  \begin{itemize}
    \item Step 6.1 of Sondovač, \texttt{sondovac$\_$part$\_$a.sh}.
    \item Depending on the phylogenetic depth that should be obtained, the probe sequences need to be designed from either the transcript or genome skim sequences, or it might not matter (if the taxa, from which the transcriptome and genome skim data were generated, are closely related).
    \item DEFAULT: no usage of -g (probe design from genome skim sequences)
    \item OPTIONS: usage of -g (probe design from transcript sequences)
  \end{itemize}
\item[\texttt{-s \#\#\#\#}] Number of BLAT hits per transcript when matching unique transcripts and the filtered, combined genome skim reads.
  \begin{itemize}
    \item Step 6.2 of Sondovač, \texttt{sondovac$\_$part$\_$a.sh}.
    \item Transcripts with a~high number of BLAT hits, indicating repetitive elements, need to be removed from the putative probe sequences.
    \item DEFAULT: 1000
    \item OPTIONS: Integer ranging from 100 to 10000
  \end{itemize}
\item[\texttt{-b \#\#\#}] Minimum exon (bait) length.
  \begin{itemize}
    \item Steps 8~and 10 of Sondovač, \texttt{sondovac$\_$part$\_$b.sh}.
    \item The minimum exon length should not fall below the bait length in order to account for specific binding between genomic libraries and baits during hybridization.
    \item DEFAULT: 120 (preferred length for phylogeny)
    \item OPTIONS: 80, 100, 120
  \end{itemize}
\item [\texttt{-k \#\#\#}] Minimum total locus length.
    \begin{itemize}
     \item Steps 8~and 10 of Sondovač, \texttt{sondovac$\_$part$\_$b.sh}.
     \item When running the script in interactive mode, the user will be asked which value to use. A~table summarizing the total number of LCN loci, which will be the result of the probe design for all minimum total locus lengths that the user can select (600 bp, 720 bp, 840 bp, 960 bp, 1080 bp, 1200 bp), will be displayed to facilitate this choice.
     \item DEFAULT: 600
     \item OPTIONS: 720, 840, 960, 1080, 1200
     \end{itemize}
\item[\texttt{-d 0.\#\#}] Sequence similarity between probe sequences (parameter -c of cd-hit-est, see its manual for details).
  \begin{itemize}
    \item Step 9~of Sondovač, \texttt{sondovac$\_$part$\_$b.sh}.
    \item Probes that target multiple similar loci need to be removed.
    \item DEFAULT: 0.9 (highly recommended)
    \item OPTIONS: Decimal ranging from 0.85 to 0.95
  \end{itemize}
\item[\texttt{-y \#\#}] Sequence similarity between probe sequences and plastome reference (parameter -minIde\-ntity of BLAT, see its manual for details).
  \begin{itemize}
    \item Step 11 of Sondovač, \texttt{sondovac$\_$part$\_$b.sh}.
    \item Some plastid reads might not have been removed in step 2; they should be removed in this step.
    \item DEFAULT: 90 (highly recommended)
    \item OPTIONS: Integer ranging from 70 to 100
  \end{itemize}
\end{description}

\subsection{Input and output files}

All names of input files and paths to them must be without spaces and without special characters (some software has difficulties handling them). \textbf{Important note:} HTS data are big. The Sondovač pipeline is relatively long, and part \texttt{A} contains several format conversions and can (for some time) use dozens of GB of disk space. Temporary files not potentially useful to the user are deleted at the end of the pipeline -- these files may be useful for debugging if something goes wrong. For example, input data of \citet{Schmickl2016} are approximately 4.5~GB, and the overall output of part \texttt{A} of the script is about 28~GB, of which less then half is kept by the pipeline. This analysis took less than an hour on an i7 3.4~GHz CPU. Part \texttt{B} is very quick and does not consume a~significant amount of disk space. All input files \textit{must} have UNIX end of lines. The script checks for it and converts the files, if needed (using \texttt{dos2unix}; typically when user runs Geneious on Windows).

\vspace{10pt}
\textbf{Script \texttt{sondovac$\_$part$\_$a.sh} requires as input files:}

\begin{enumerate}
  \item Transcriptome input file in FASTA format. \textbf{Note:} For technical reasons, the labels of FASTA sequences \textit{must} be unique numbers (no other characters). Sondovač will check the labels, and if they are not in an appropriate form, a~copy of this input file with correct labels will be created.
  \item Plastome reference sequence input file in FASTA format.
  \item Paired-end genome skim input file in FASTQ format (two files -- forward and reverse reads).
  \item OPTIONAL: Mitochondriome reference sequence input file in FASTA format. This file is not required.
\end{enumerate}

\textbf{Script \texttt{sondovac$\_$part$\_$a.sh} creates the following files:}
\label{partb}

\begin{enumerate}
  \item \texttt{*$\_$renamed.fasta} -- A~copy of the transcriptome input file with the changed labels of the FASTA sequences (unique numbers corresponding to the line numbers in the original file). File \texttt{*$\_$old$\_$and$\_$new$\_$names.tsv} then contains two columns: \textbf{1)} the original sequence labels as in the user-provided transcriptome input file and \textbf{2)} new sequence labels. This might be useful to trace back certain sequences/probes.
  \item \texttt{*$\_$blat$\_$unique$\_$transcripts.psl} -- Output of BLAT (removal of transcripts sharing $\geq$90\% sequence similarity).
  \item \texttt{*$\_$unique$\_$transcripts.fasta} -- Unique transcripts in FASTA format.
  \item \texttt{*$\_$genome$\_$skim$\_$data$\_$no$\_$cp$\_$reads} -- Genome skim data without cpDNA reads.
  \item \texttt{*$\_$genome$\_$skim$\_$data$\_$no$\_$cp$\_$no$\_$mt$\_$reads} -- Genome skim data without mtDNA reads -- only if mitochondriome reference sequence was used.
  \item \texttt{*$\_$combined$\_$reads$\_$co$\_$cp$\_$no$\_$mt$\_$reads} -- Combined paired-end genome skim reads.
  \item \texttt{*$\_$blat$\_$unique$\_$transcripts$\_$versus$\_$genome$\_$skim$\_$data.pslx} -- Output of BLAT (ma\-tching of the unique transcripts and the filtered, combined genome skim reads sharing $\geq$85\% sequence similarity).
  \item \texttt{*$\_$blat$\_$unique$\_$transcripts$\_$versus$\_$genome$\_$skim$\_$data.fasta} -- Matching sequences in FASTA.
  \item \texttt{*$\_$blat$\_$unique$\_$transcripts$\_$versus$\_$genome$\_$skim$\_$data-no$\_$missing$\_$fin.fsa} -- \textbf{Part A, final FASTA sequences for usage in Geneious} (step 7, see chapter~\ref{geneious-usage} at page~\pageref{geneious-usage}, and page~\pageref{pipeline-overview}).
\end{enumerate}

Files 1-8 are not necessary for further processing by this pipeline, but may be useful to the user. The last file (9) is used as input file for Geneious in the next step. An asterisk (*) denotes the beginning of the output files' names specified by the user with parameter \texttt{-o}. If the user does not select a~custom name, default value (\texttt{output}) will be used.

\textbf{Geneious} requires as input the last output file of \texttt{sondovac$\_$part$\_$a.sh} (file 9: \texttt{*$\_$blat$\_$uni\- que$\_$transcripts$\_$versus$\_$genome$\_$skim$\_$data-no$\_$missing$\_$fin.fsa}). The output from Gene\-ious consists of two files (see page~\pageref{geneious-usage}):

\begin{enumerate}
  \item Final assembled sequences exported as TSV.
  \item Final assembled sequences exported as FASTA.
\end{enumerate}

\textbf{Script \texttt{sondovac$\_$part$\_$b.sh} requires as input files:}

\begin{enumerate}
  \item Plastome reference sequence input file in FASTA format.
  \item Assembled sequences exported from Geneious as TSV.
  \item Assembled sequences exported from Geneious as FASTA.
\end{enumerate}

\textbf{Script \texttt{sondovac$\_$part$\_$b.sh} creates the following files:}

\begin{enumerate}
  \item \texttt{*$\_$prelim$\_$probe$\_$seq.fasta} -- Preliminary probe sequences.
  \item \texttt{*$\_$prelim$\_$probe$\_$seq$\_$cluster$\_$100.fasta} -- Unclustered exons and clustered exons with 100\% sequence identity.
  \item \texttt{*$\_$prelim$\_$probe$\_$seq$\_$cluster$\_$90.clstr} -- Unclustered exons and clustered exons with more than a certain sequence similarity (CLSTR file).
  \item \texttt{*$\_$unique$\_$prelim$\_$probe$\_$seq.fasta} -- Unclustered exons / exons with less than a certain sequence similarity.
  \item \texttt{*$\_$similarity$\_$test.fasta} -- Contigs that comprise exons $\geq$ bait length and have a~certain total locus length.
  \item \texttt{*$\_$target$\_$enrichment$\_$probe$\_$sequences$\_$with$\_$pt.fasta} -- All probes in FASTA, with putative plastid sequences (if there were any BLAT hits, putative plastid sequences are listed in next file).
  \item \texttt{*$\_$possible$\_$cp$\_$dna$\_$gene$\_$in$\_$probe$\_$set.pslx} -- In case of any BLAT hits, putative remaining plastid probe sequences from \texttt{*$\_$target$\_$enrichm\-ent$\_$probe$\_$sequences$\_$with$\_$pt\-.fasta} are listed here. \textbf{Not removing plastid genes will take lots of space on the Illumina lane for enriched plastid reads that should actually be available for enriched nuclear reads.}
  \item \underline{\texttt{*$\_$target$\_$enrichment$\_$probe$\_$sequences.fasta}} -- \textbf{Final probes in FASTA.}
\end{enumerate}

An asterisk (\texttt{*}) denotes the beginning of the output files' names specified by the user with parameter \texttt{-o}. If the user does not select a~custom name, the default value (\texttt{output}) will be used. By default, output files are created in the same directory from which Sondovač was launched. Output files can be saved in a~custom directory by specifying an output directory with parameter \texttt{-o}:

\begin{bashcode}
  # Find current directory (e.g. /home/user):
  pwd
  # Launching Sondovač located in directory /home/user/sondovac
  # and save output to e.g. desktop (/home/user/Desktop):
  ./sondovac/sondovac_part_a.sh -o Desktop/MyFile
  # Sondovač will save software (if needed) in "bin" directory
  # located in directory from which it was launched, see it:
  ls bin/*
  # Output files are in desired directory, see them e.g. by:
  ls -lh Desktop/MyFile*
\end{bashcode}

\subsection{Geneious usage}
\label{geneious-usage}

Import the output file of part~A~of the script (\texttt{sondovac$\_$part$\_$a.sh}): go to menu \textbf{File | Import | From File\ldots} This file is named as: \texttt{*$\_$blat$\_$unique$\_$transcripts$\_$versus$\_$genome$\_$skim$\_$data\--no$\_$missing$\_$fin.fsa} (see Figure~\ref{geneious-import}).

\begin{figure}[hbt]
  \includegraphics[width=\textwidth]{geneious1.png}
  \caption[Import into Geneious]{Import of output of \texttt{sondovac$\_$part$\_$a.sh} into Geneious for next processing (see page~\pageref{pipeline-overview}). Go to menu File \textbf{(1)} | Import \textbf{(2)} | From file\ldots \textbf{(3)} and import output of \texttt{sondovac$\_$part$\_$a.sh}. You should see a~similar result as \textbf{(4)}.}
  \label{geneious-import}
\end{figure}

Select the file and go to menu \textbf{Tools | Align / Assemble | De Novo Assemble}. In \textbf{Data} frame select \textbf{Assemble by 1st (\ldots) Underscore}. In \textbf{Method} frame select \textbf{Geneious Assembler} (if you don't have other assemblers, this option might be missing) and \textbf{Medium Sensitivity / Fast} sensitivity (see Figure~\ref{geneious-assembly}).

In \textbf{Results} frame check \textbf{Save assembly report}, \textbf{Save list of unused reads}, \textbf{Save in sub-folder}, \textbf{Save contigs} (do not check \textbf{Maximum}) and \textbf{Save consensus sequences} (Click to \textit{Options} -- \textbf{Save consensus used by assembler} must be selected.). \textbf{Do not trim}. Otherwise keep defaults. Run it. Geneious may warn about possible hanging because of big file size. Do not use Geneious for other tasks during the assembly. Running Geneious may take a~long time (see Figure~\ref{geneious-assembly}).

\begin{figure}[htb]
  \begin{center}
    \includegraphics[width=12cm]{geneious2.png}
  \end{center}
  \caption[Settings of Geneious assembly]{Settings of Geneious assembly as described in the main text. It can take a~longer time to run it.}
  \label{geneious-assembly}
\end{figure}

Select all resulting contigs (typically named \textbf{* Contig \#}) and export them (go to menu \textbf{File | Export | Selected Documents\ldots}) as \textbf{Tab-separated table values (*.tsv)}. Save the following columns (Hold \texttt{Ctrl} key to mark more fields): \textbf{\# Sequences}, \textbf{\% Pairwise Identity}, \textbf{Description}, \textbf{Mean Coverage}, \textbf{Name} and \textbf{Sequence Length}. If this option is inaccessible to you, export all columns (see Figure~\ref{geneious-export1}). Warning! Do not select and export \textbf{* Consensus Sequences}, \textbf{* Unused Reads} or \textbf{* Report} -- only the individual \textbf{* contig \#} files (see Figure~\ref{geneious-export1}).

\begin{figure}[p]
  \includegraphics[width=\textwidth]{geneious3.png}
  \caption[Export of contigs as TSV from Geneious]{Select all (and only) \textbf{* Contig \#} files \textbf{(1)}. Go to menu File \textbf{(2)} | Export \textbf{(3)} | Selected Documents\ldots \textbf{(4)} and export them as Tab-separated table values (TSV) \textbf{(5)}. Export only marked columns \textbf{(6)} (hold \texttt{Ctrl} to mark more fields).}
  \label{geneious-export1}
\end{figure}

Select items \textbf{Consensus Sequences} and \textbf{Unused Reads} and export them as one \textbf{FASTA}. Go to menu \textbf{File | Export | Selected Documents\ldots} and choose \textbf{FASTA file type} (see Figure~\ref{geneious-export2}).

\begin{figure}[htb]
  \includegraphics[width=\textwidth]{geneious4.png}
  \caption[Export of FASTA from Geneious]{ Select only documents \textbf{Consensus Sequences} and \textbf{Unused Reads} and export them as FASTA format (see also Figure~\ref{geneious-export1}).}
  \label{geneious-export2}
\end{figure}

Use the exported files from Geneious as input for part~B of the script (\texttt{sondovac$\_$part$\_$b.sh}).

\subsection{Record output of Sondovač}

To record the whole output of Sondovač (regardless of used parameters), use utility \texttt{tee}. This will produce a~plain text output with everything printed to the screen. It can be useful for reference or exploration if something went wrong. Use it as follows:

\begin{bashcode}
  ./sondovac_part_a.sh | tee records.log
  man tee # See more options how tee can record the script's output
  # "|" is a~pipe passing output of the 1st command as input for the 2nd command
  less records.log # See the record. Quit viewing by "Q"
  rm records.log # Delete the log file
\end{bashcode}

You can use any command line arguments; the script will behave as usual. The plain text file \texttt{records.log} will then contain all its output. Unfortunately, \texttt{tee} usually wrongly records "invisible" characters -- tabs and coloration used to highlight user messages in the script. If you see weird characters in texttt{records.log} that disturb reading, use the following commands:

\begin{bashcode}
  # Assume output of Sondovač is named "records.log"
  sed -i 's/.\[[[:digit:]]\{1,2\}m//g' records.log
  # Explanation of regular expression (find pattern and replace by nothing):
  # any character, [, one/two number(s), m (sequence defining text formatting)
  sed -i 's/.(B.\[m//g' records.log
  # Explanation of regular expression (find pattern and replace by nothing):
  # any character, (, B, [, m (sequence defining text formatting)
  # Escaping \[ \] is required to search specifically for brackets []
  # (NOT searching for any character within [...] - there is no escaping)
  # but \{...\} define number of occurrences of previous character(s)
\end{bashcode}

Note for Mac OS~X users: the regular expressions above require GNU \texttt{sed}, not the version presented by default in Mac OS~X. It is installed by Sondovač using Homebrew (see page~\pageref{required-mac}), but to launch it you probably must use the command \texttt{gsed} instead of \texttt{sed}.

\section{Sample data}

Together with the script, we provide the ZIP archive (1.8 GB) that contains example input files for running the script: \textit{Oxalis} genome skim data and transcriptome as well as the \textit{Ricinus} cpDNA and mtDNA reference sequences. See \url{https://github.com/V-Z/sondovac/wiki/Sample-data} for download of sample data.

The package contains:

\begin{enumerate}
  \item \texttt{input1$\_$JHCN$\_$Oxalis$\_$corniculata$\_$transcriptome$\_$data.fa} -- \textit{Oxalis corniculata} transcriptome (parameter \texttt{-f}).
  \item \texttt{input2$\_$Ricinus$\_$communis$\_$reference$\_$plastid$\_$genome.fsa} -- cpDNA reference (para\-meter \texttt{-c}), GenBank reference number \href{https://www.ncbi.nlm.nih.gov/nuccore/372450118/}{NC$\_$016736}.
  \item \texttt{input3$\_$J12$\_$Oxalis$\_$obtusa$\_$J12$\_$genome$\_$skim$\_$data$\_$R1.fastq} -- paired-end genome sk\-im data, file 1~(forward reads, parameter \texttt{-t}).
  \item \texttt{input4$\_$J12$\_$Oxalis$\_$obtusa$\_$J12$\_$genome$\_$skim$\_$data$\_$R2.fastq} -- paired-end genome sk\-im data, file 2~(backward reads, parameter \texttt{-q}).
  \item \texttt{input5$\_$Ricinus$\_$communis$\_$reference$\_$mitochondrial$\_$genome.fasta} -- mtDNA reference (parameter \texttt{-m}), GenBank reference number \href{https://www.ncbi.nlm.nih.gov/nuccore/323649872/}{NC$\_$015141}.
\end{enumerate}

The transcriptome input file is unpublished data from G. K.-S. Wong et al. Data can be also found under

\begin{itemize}
  \item \url{http://www.onekp.com/}
  \item \url{http://www.onekp.com/samples/list.php}
  \item \url{http://www.onekp.com/samples/single.php?id=JHCN}
\end{itemize}

The transcriptome FASTA file used for the probe design is named \texttt{JHCN-SOAPdenovo-Trans-assembly.dnas.out} and can be found under \texttt{JHCN/Assembly/JHCN-SOAPdenovo-Trans-translated/}. Information about how to get access to data download is given in \citet{Matasci2014}.

Explanation of command line parameters is on page~\pageref{script-usage}. Download all data from \url{https://github.com/V-Z/sondovac/releases/tag/sample-data-2.0}.

\section{Questions not covered here, reporting bugs and wishes}

If you have a question or encounter a problem, please see \url{https://github.com/V-Z/sondovac/issues} and feel free to ask any question and/or express any wish. The authors will do their best to help you.

\section{Changelog} % TODO Update changelog

List of changes in released versions of Sondovač.

\subsection{Version 1.3 regular release released 2017-MM-DD}

\begin{itemize}
  \item \texttt{bam2fastq} is dropped in favour of \texttt{samtools fastq}. No plans to use \texttt{Picard} anymore (part~A).
  \item Simplified \texttt{INSTALL} and \texttt{README} not to only copy PDF manual.
  \item Corrected output of part~A -- ensure to have always valid FASTA.
  \item Automatically remove putative plastid sequences from final probe set (part~B), list of all probes (with putative plastid sequences) and list of putative plastid sequences are available.
  \item Updated software distributed with Sondovač, updated respective sections of manual.
  \item Removed FASTX Toolkit, conversion from FASTQ to FASTA is done by simple shell function.
\end{itemize}

\subsection{Version 1.2 regular release released 2016-06-28}

\begin{itemize}
  \item Fixed wrong probe design summary statistics (at the end of part~B).
  \item Added support for Geneious 9.
  \item More detailed error reporting (especially in part~B).
  \item Various smaller fixes.
  \item Updated accompanying software to recent versions.
\end{itemize}

\subsection{Version 1.0 regular release released 2016-01-12}

\begin{itemize}
  \item Renaming of input FASTA sequences names is required - it ensures correct working of part~B.
  \item Added check if input files were created on Windows - if so, they are converted into UNIX style EOL.
  \item Various smaller fixes.
  \item Better showing of the information in part~B.
  \item Enhanced documentation.
\end{itemize}

\subsection{Version 0.99 release candidate released 2015-12-08}

\begin{itemize}
  \item Fixed error with some input files for part~B.
  \item Finished coloration of command-line user interface.
  \item Added possibility to set minimal exon length of the loci.
  \item Various fixes and UI enhancements.
  \item Improved documentation.
\end{itemize}

\subsection{Version 0.95 beta released 2015-11-27}

\begin{itemize}
  \item Offer the possibility to choose between transcripts or genome skim sequences for further processing in step 6.1, part~A.
  \item Coloration of command-line user interface (incomplete).
  \item Added possibility to change -minIdentity parameter of BLAT in step 11, part~B.
  \item Fixed problems with some transcriptome input files.
  \item Added possibility to set custom bait length.
  \item Added information about article in MER introducing Sondovač.
\end{itemize}

\subsection{Version 0.9 beta released 2015-10-23}

\begin{itemize}
  \item Highly enhanced part~B.
  \item Better handling of variable output from Geneious.
  \item Possibility to specify the name of the custom output file.
  \item Full support for Linux distributions using DEB -- Debian, Ubuntu, Linux Mint and derivatives.
  \item Enhanced documentation.
  \item Support for Mac OS X, package management using Homebrew.
  \item Support for RedHat based Linux distributions -- Fedora, Centos and Scientific Linux and derivatives.
  \item Better compilation and installation of required software.
  \item For downloading automatically select whether to use wget (preferred) or curl.
  \item Various fixes.
\end{itemize}

\subsection{Version 0.8 alpha released 2015-10-09}

\begin{itemize}
  \item Usage of mitochondrial reference sequence is optional.
  \item Better formatting of script messages.
  \item Various fixes and enhancements.
\end{itemize}

\subsection{Version 0.7 alpha released 2015-10-06}

\begin{itemize}
  \item Fixed reported problems with sed differences among Linux and Mac OS X.
  \item Added more exhaustive documentation.
  \item Various fixes and enhancements.
\end{itemize}

\subsection{Version 0.6 alpha released 2015-08-10}

\begin{itemize}
  \item Fixed problems with some versions of output of Geneious.
  \item Better compilation and installation of required additional software packages.
  \item Various fixes and enhancements.
\end{itemize}

\subsection{Version 0.5 alpha released 2015-07-24}

\begin{itemize}
  \item First public release, early alpha stage.
\end{itemize}

\section{Licenses}

The set of BASH scripts Sondovač is licensed under GNU General Public License version 3. List of licenses of included software is in Table~\ref{software-lic} (see full texts below). License of BLAT does not allow redistribution, so that this software is not included and the software is downloaded on the fly. Script is also using software included in GNU core utilities (basic tools available in any UNIX-based system), see \url{https://www.gnu.org/software/coreutils/} for details.

\begin{longtable}{| >{\centering\arraybackslash}m{2.5cm} >{\centering\arraybackslash}m{3.4cm} >{\centering\arraybackslash}m{8.8cm} |}
\caption[List of software and licenses]{List of software, licenses and links to license details.}\\
\hline
\endhead
\hline
\endfoot
\textbf{Software} & \textbf{License} & \textbf{License details}\\
Sondovač & GNU GPL v. 3 & \url{https://gnu.org/licenses/gpl.html}\\
Bowtie2 & GNU GPL v. 3 & \url{https://gnu.org/licenses/gpl.html}\\
SAMtools & MIT/Expat License & \url{https://en.wikipedia.org/wiki/MIT_License}\\
FLASh & GNU GPL v. 3 & \url{https://gnu.org/licenses/gpl.html}\\
CD-HIT & GNU GPL v. 2 & \url{https://gnu.org/licenses/old-licenses/gpl-2.0.html}\\
grap$\_$synglet\-on$\_$clusters.py & GNU GPL v. 3 & \url{https://gnu.org/licenses/gpl.html}
\label{software-lic}
\end{longtable}

\begingroup
\fontsize{7pt}{8pt}
\selectfont

\subsection{GNU General Public License, Version 3, 29 June 2007}

Copyright © 2007 Free Software Foundation, Inc. \url{https://fsf.org/}

Everyone is permitted to copy and distribute verbatim copies of this license document, but changing it is not allowed.

\subsubsection{Preamble}

The GNU General Public License is a~free, copyleft license for software and other kinds of works.

The licenses for most software and other practical works are designed to take away your freedom to share and change the works. By contrast, the GNU General Public License is intended to guarantee your freedom to share and change all versions of a~program--to make sure it remains free software for all its users. We, the Free Software Foundation, use the GNU General Public License for most of our software; it applies also to any other work released this way by its authors. You can apply it to your programs, too.

When we speak of free software, we are referring to freedom, not price. Our General Public Licenses are designed to make sure that you have the freedom to distribute copies of free software (and charge for them if you wish), that you receive source code or can get it if you want it, that you can change the software or use pieces of it in new free programs, and that you know you can do these things.

To protect your rights, we need to prevent others from denying you these rights or asking you to surrender the rights. Therefore, you have certain responsibilities if you distribute copies of the software, or if you modify it: responsibilities to respect the freedom of others.

For example, if you distribute copies of such a~program, whether gratis or for a~fee, you must pass on to the recipients the same freedoms that you received. You must make sure that they, too, receive or can get the source code. And you must show them these terms so they know their rights.

Developers that use the GNU GPL protect your rights with two steps: (1) assert copyright on the software, and (2) offer you this License giving you legal permission to copy, distribute and/or modify it.

For the developers' and authors' protection, the GPL clearly explains that there is no warranty for this free software. For both users' and authors' sake, the GPL requires that modified versions be marked as changed, so that their problems will not be attributed erroneously to authors of previous versions.

Some devices are designed to deny users access to install or run modified versions of the software inside them, although the manufacturer can do so. This is fundamentally incompatible with the aim of protecting users' freedom to change the software. The systematic pattern of such abuse occurs in the area of products for individuals to use, which is precisely where it is most unacceptable. Therefore, we have designed this version of the GPL to prohibit the practice for those products. If such problems arise substantially in other domains, we stand ready to extend this provision to those domains in future versions of the GPL, as needed to protect the freedom of users.

Finally, every program is threatened constantly by software patents. States should not allow patents to restrict development and use of software on general-purpose computers, but in those that do, we wish to avoid the special danger that patents applied to a~free program could make it effectively proprietary. To prevent this, the GPL assures that patents cannot be used to render the program non-free.

The precise terms and conditions for copying, distribution and modification follow.

\subsubsection{Terms and Conditions}

\subsubsection{0. Definitions}

“This License” refers to version 3~of the GNU General Public License.

“Copyright” also means copyright-like laws that apply to other kinds of works, such as semiconductor masks.

“The Program” refers to any copyrightable work licensed under this License. Each licensee is addressed as “you”. “Licensees” and “recipients” may be individuals or organizations.

To “modify” a~work means to copy from or adapt all or part of the work in a~fashion requiring copyright permission, other than the making of an exact copy. The resulting work is called a~“modified version” of the earlier work or a~work “based on” the earlier work.

A~“covered work” means either the unmodified Program or a~work based on the Program.

To “propagate” a~work means to do anything with it that, without permission, would make you directly or secondarily liable for infringement under applicable copyright law, except executing it on a~computer or modifying a~private copy. Propagation includes copying, distribution (with or without modification), making available to the public, and in some countries other activities as well.

To “convey” a~work means any kind of propagation that enables other parties to make or receive copies. Mere interaction with a~user through a~computer network, with no transfer of a~copy, is not conveying.

An interactive user interface displays “Appropriate Legal Notices” to the extent that it includes a~convenient and prominently visible feature that (1) displays an appropriate copyright notice, and (2) tells the user that there is no warranty for the work (except to the extent that warranties are provided), that licensees may convey the work under this License, and how to view a~copy of this License. If the interface presents a~list of user commands or options, such as a~menu, a~prominent item in the list meets this criterion.

\subsubsection{1. Source Code}

The “source code” for a~work means the preferred form of the work for making modifications to it. “Object code” means any non-source form of a~work.

A~“Standard Interface” means an interface that either is an official standard defined by a~recognized standards body, or, in the case of interfaces specified for a~particular programming language, one that is widely used among developers working in that language.

The “System Libraries” of an executable work include anything, other than the work as a~whole, that (a) is included in the normal form of packaging a~Major Component, but which is not part of that Major Component, and (b) serves only to enable use of the work with that Major Component, or to implement a~Standard Interface for which an implementation is available to the public in source code form. A~“Major Component”, in this context, means a~major essential component (kernel, window system, and so on) of the specific operating system (if any) on which the executable work runs, or a~compiler used to produce the work, or an object code interpreter used to run it.

The “Corresponding Source” for a~work in object code form means all the source code needed to generate, install, and (for an executable work) run the object code and to modify the work, including scripts to control those activities. However, it does not include the work's System Libraries, or general-purpose tools or generally available free programs which are used unmodified in performing those activities but which are not part of the work. For example, Corresponding Source includes interface definition files associated with source files for the work, and the source code for shared libraries and dynamically linked subprograms that the work is specifically designed to require, such as by intimate data communication or control flow between those subprograms and other parts of the work.

The Corresponding Source need not include anything that users can regenerate automatically from other parts of the Corresponding Source.

The Corresponding Source for a~work in source code form is that same work.

\subsubsection{2. Basic Permissions}

All rights granted under this License are granted for the term of copyright on the Program, and are irrevocable provided the stated conditions are met. This License explicitly affirms your unlimited permission to run the unmodified Program. The output from running a~covered work is covered by this License only if the output, given its content, constitutes a~covered work. This License acknowledges your rights of fair use or other equivalent, as provided by copyright law.

You may make, run and propagate covered works that you do not convey, without conditions so long as your license otherwise remains in force. You may convey covered works to others for the sole purpose of having them make modifications exclusively for you, or provide you with facilities for running those works, provided that you comply with the terms of this License in conveying all material for which you do not control copyright. Those thus making or running the covered works for you must do so exclusively on your behalf, under your direction and control, on terms that prohibit them from making any copies of your copyrighted material outside their relationship with you.

Conveying under any other circumstances is permitted solely under the conditions stated below. Sublicensing is not allowed; section 10 makes it unnecessary.

\subsubsection{3. Protecting Users' Legal Rights From Anti-Circumvention Law}

No covered work shall be deemed part of an effective technological measure under any applicable law fulfilling obligations under article 11 of the WIPO copyright treaty adopted on 20 December 1996, or similar laws prohibiting or restricting circumvention of such measures.

When you convey a~covered work, you waive any legal power to forbid circumvention of technological measures to the extent such circumvention is effected by exercising rights under this License with respect to the covered work, and you disclaim any intention to limit operation or modification of the work as a~means of enforcing, against the work's users, your or third parties' legal rights to forbid circumvention of technological measures.

\subsubsection{4. Conveying Verbatim Copies}

You may convey verbatim copies of the Program's source code as you receive it, in any medium, provided that you conspicuously and appropriately publish on each copy an appropriate copyright notice; keep intact all notices stating that this License and any non-permissive terms added in accord with section 7~apply to the code; keep intact all notices of the absence of any warranty; and give all recipients a~copy of this License along with the Program.

You may charge any price or no price for each copy that you convey, and you may offer support or warranty protection for a~fee.

\subsubsection{5. Conveying Modified Source Versions}

You may convey a~work based on the Program, or the modifications to produce it from the Program, in the form of source code under the terms of section 4, provided that you also meet all of these conditions:

\begin{enumerate}[label=\Alph*)]
\item The work must carry prominent notices stating that you modified it, and giving a~relevant date.
\item The work must carry prominent notices stating that it is released under this License and any conditions added under section 7. This requirement modifies the requirement in section 4~to “keep intact all notices”.
\item You must license the entire work, as a~whole, under this License to anyone who comes into possession of a~copy. This License will therefore apply, along with any applicable section 7~additional terms, to the whole of the work, and all its parts, regardless of how they are packaged. This License gives no permission to license the work in any other way, but it does not invalidate such permission if you have separately received it.
\item If the work has interactive user interfaces, each must display Appropriate Legal Notices; however, if the Program has interactive interfaces that do not display Appropriate Legal Notices, your work need not make them do so.
\end{enumerate}

A~compilation of a~covered work with other separate and independent works, which are not by their nature extensions of the covered work, and which are not combined with it such as to form a~larger program, in or on a~volume of a~storage or distribution medium, is called an “aggregate” if the compilation and its resulting copyright are not used to limit the access or legal rights of the compilation's users beyond what the individual works permit. Inclusion of a~covered work in an aggregate does not cause this License to apply to the other parts of the aggregate.

\subsubsection{6. Conveying Non-Source Forms}

You may convey a~covered work in object code form under the terms of sections 4~and 5, provided that you also convey the machine-readable Corresponding Source under the terms of this License, in one of these ways:

\begin{enumerate}[label=\Alph*)]
\item Convey the object code in, or embodied in, a~physical product (including a~physical distribution medium), accompanied by the Corresponding Source fixed on a~durable physical medium customarily used for software interchange.
\item Convey the object code in, or embodied in, a~physical product (including a~physical distribution medium), accompanied by a~written offer, valid for at least three years and valid for as long as you offer spare parts or customer support for that product model, to give anyone who possesses the object code either (1) a~copy of the Corresponding Source for all the software in the product that is covered by this License, on a~durable physical medium customarily used for software interchange, for a~price no more than your reasonable cost of physically performing this conveying of source, or (2) access to copy the Corresponding Source from a~network server at no charge.
\item Convey individual copies of the object code with a~copy of the written offer to provide the Corresponding Source. This alternative is allowed only occasionally and noncommercially, and only if you received the object code with such an offer, in accord with subsection 6b.
\item Convey the object code by offering access from a~designated place (gratis or for a~charge), and offer equivalent access to the Corresponding Source in the same way through the same place at no further charge. You need not require recipients to copy the Corresponding Source along with the object code. If the place to copy the object code is a~network server, the Corresponding Source may be on a~different server (operated by you or a~third party) that supports equivalent copying facilities, provided you maintain clear directions next to the object code saying where to find the Corresponding Source. Regardless of what server hosts the Corresponding Source, you remain obligated to ensure that it is available for as long as needed to satisfy these requirements.
\item Convey the object code using peer-to-peer transmission, provided you inform other peers where the object code and Corresponding Source of the work are being offered to the general public at no charge under subsection 6d.
\end{enumerate}

A~separable portion of the object code, whose source code is excluded from the Corresponding Source as a~System Library, need not be included in conveying the object code work.

A~“User Product” is either (1) a~“consumer product”, which means any tangible personal property which is normally used for personal, family, or household purposes, or (2) anything designed or sold for incorporation into a~dwelling. In determining whether a~product is a~consumer product, doubtful cases shall be resolved in favor of coverage. For a~particular product received by a~particular user, “normally used” refers to a~typical or common use of that class of product, regardless of the status of the particular user or of the way in which the particular user actually uses, or expects or is expected to use, the product. A~product is a~consumer product regardless of whether the product has substantial commercial, industrial or non-consumer uses, unless such uses represent the only significant mode of use of the product.

“Installation Information” for a~User Product means any methods, procedures, authorization keys, or other information required to install and execute modified versions of a~covered work in that User Product from a~modified version of its Corresponding Source. The information must suffice to ensure that the continued functioning of the modified object code is in no case prevented or interfered with solely because modification has been made.

If you convey an object code work under this section in, or with, or specifically for use in, a~User Product, and the conveying occurs as part of a~transaction in which the right of possession and use of the User Product is transferred to the recipient in perpetuity or for a~fixed term (regardless of how the transaction is characterized), the Corresponding Source conveyed under this section must be accompanied by the Installation Information. But this requirement does not apply if neither you nor any third party retains the ability to install modified object code on the User Product (for example, the work has been installed in ROM).

The requirement to provide Installation Information does not include a~requirement to continue to provide support service, warranty, or updates for a~work that has been modified or installed by the recipient, or for the User Product in which it has been modified or installed. Access to a~network may be denied when the modification itself materially and adversely affects the operation of the network or violates the rules and protocols for communication across the network.

Corresponding Source conveyed, and Installation Information provided, in accord with this section must be in a~format that is publicly documented (and with an implementation available to the public in source code form), and must require no special password or key for unpacking, reading or copying.

\subsubsection{7. Additional Terms}

“Additional permissions” are terms that supplement the terms of this License by making exceptions from one or more of its conditions. Additional permissions that are applicable to the entire Program shall be treated as though they were included in this License, to the extent that they are valid under applicable law. If additional permissions apply only to part of the Program, that part may be used separately under those permissions, but the entire Program remains governed by this License without regard to the additional permissions.

When you convey a~copy of a~covered work, you may at your option remove any additional permissions from that copy, or from any part of it. (Additional permissions may be written to require their own removal in certain cases when you modify the work.) You may place additional permissions on material, added by you to a~covered work, for which you have or can give appropriate copyright permission.

Notwithstanding any other provision of this License, for material you add to a~covered work, you may (if authorized by the copyright holders of that material) supplement the terms of this License with terms:

\begin{enumerate}[label=\Alph*)]
\item Disclaiming warranty or limiting liability differently from the terms of sections 15 and 16 of this License; or
\item Requiring preservation of specified reasonable legal notices or author attributions in that material or in the Appropriate Legal Notices displayed by works containing it; or
\item Prohibiting misrepresentation of the origin of that material, or requiring that modified versions of such material be marked in reasonable ways as different from the original version; or
\item Limiting the use for publicity purposes of names of licensors or authors of the material; or
\item Declining to grant rights under trademark law for use of some trade names, trademarks, or service marks; or 
\item Requiring indemnification of licensors and authors of that material by anyone who conveys the material (or modified versions of it) with contractual assumptions of liability to the recipient, for any liability that these contractual assumptions directly impose on those licensors and authors.
\end{enumerate}

All other non-permissive additional terms are considered “further restrictions” within the meaning of section 10. If the Program as you received it, or any part of it, contains a~notice stating that it is governed by this License along with a~term that is a~further restriction, you may remove that term. If a~license document contains a~further restriction but permits relicensing or conveying under this License, you may add to a~covered work material governed by the terms of that license document, provided that the further restriction does not survive such relicensing or conveying.

If you add terms to a~covered work in accord with this section, you must place, in the relevant source files, a~statement of the additional terms that apply to those files, or a~notice indicating where to find the applicable terms.

Additional terms, permissive or non-permissive, may be stated in the form of a~separately written license, or stated as exceptions; the above requirements apply either way.

\subsubsection{8. Termination}

You may not propagate or modify a~covered work except as expressly provided under this License. Any attempt otherwise to propagate or modify it is void, and will automatically terminate your rights under this License (including any patent licenses granted under the third paragraph of section 11).

However, if you cease all violation of this License, then your license from a~particular copyright holder is reinstated (a) provisionally, unless and until the copyright holder explicitly and finally terminates your license, and (b) permanently, if the copyright holder fails to notify you of the violation by some reasonable means prior to 60 days after the cessation.

Moreover, your license from a~particular copyright holder is reinstated permanently if the copyright holder notifies you of the violation by some reasonable means, this is the first time you have received notice of violation of this License (for any work) from that copyright holder, and you cure the violation prior to 30 days after your receipt of the notice.

Termination of your rights under this section does not terminate the licenses of parties who have received copies or rights from you under this License. If your rights have been terminated and not permanently reinstated, you do not qualify to receive new licenses for the same material under section 10.

\subsubsection{9. Acceptance Not Required for Having Copies}

You are not required to accept this License in order to receive or run a~copy of the Program. Ancillary propagation of a~covered work occurring solely as a~consequence of using peer-to-peer transmission to receive a~copy likewise does not require acceptance. However, nothing other than this License grants you permission to propagate or modify any covered work. These actions infringe copyright if you do not accept this License. Therefore, by modifying or propagating a~covered work, you indicate your acceptance of this License to do
so.

\subsubsection{10. Automatic Licensing of Downstream Recipients}

Each time you convey a~covered work, the recipient automatically receives a~license from the original licensors, to run, modify and propagate that work, subject to this License. You are not responsible for enforcing compliance by third parties with this License.

An “entity transaction” is a~transaction transferring control of an organization, or substantially all assets of one, or subdividing an organization, or merging organizations. If propagation of a~covered work results from an entity transaction, each party to that transaction who receives a~copy of the work also receives whatever licenses to the work the party's predecessor in interest had or could give under the previous paragraph, plus a~right to possession of the Corresponding Source of the work from the predecessor in interest, if the predecessor has it or can get it with reasonable efforts.

You may not impose any further restrictions on the exercise of the rights granted or affirmed under this License. For example, you may not impose a~license fee, royalty, or other charge for exercise of rights granted under this License, and you may not initiate litigation (including a~cross-claim or counterclaim in a~lawsuit) alleging that any patent claim is infringed by making, using, selling, offering for sale, or importing the Program or any portion of it.

\subsubsection{11. Patents}

A~“contributor” is a~copyright holder who authorizes use under this License of the Program or a~work on which the Program is based. The work thus licensed is called the contributor's “contributor version”.

A~contributor's “essential patent claims” are all patent claims owned or controlled by the contributor, whether already acquired or hereafter acquired, that would be infringed by some manner, permitted by this License, of making, using, or selling its contributor version, but do not include claims that would be infringed only as a~consequence of further modification of the contributor version. For purposes of this definition, “control” includes the right to grant patent sublicenses in a~manner consistent with the requirements of this License.

Each contributor grants you a~non-exclusive, worldwide, royalty-free patent license under the contributor's essential patent claims, to make, use, sell, offer for sale, import and otherwise run, modify and propagate the contents of its contributor version.

In the following three paragraphs, a~“patent license” is any express agreement or commitment, however denominated, not to enforce a~patent (such as an express permission to practice a~patent or covenant not to sue for patent infringement). To “grant” such a~patent license to a~party means to make such an agreement or commitment not to enforce a~patent against the party.

If you convey a~covered work, knowingly relying on a~patent license, and the Corresponding Source of the work is not available for anyone to copy, free of charge and under the terms of this License, through a~publicly available network server or other readily accessible means, then you must either (1) cause the Corresponding Source to be so available, or (2) arrange to deprive yourself of the benefit of the patent license for this particular work, or (3) arrange, in a~manner consistent with the requirements of this License, to extend the patent license to downstream recipients. “Knowingly relying” means you have actual knowledge that, but for the patent license, your conveying the covered work in a~country, or your recipient's use of the covered work in a~country, would infringe one or more identifiable patents in that country that you have reason to believe are valid.

If, pursuant to or in connection with a~single transaction or arrangement, you convey, or propagate by procuring conveyance of, a~covered work, and grant a~patent license to some of the parties receiving the covered work authorizing them to use, propagate, modify or convey a~specific copy of the covered work, then the patent license you grant is automatically extended to all recipients of the covered work and works based on it.

A~patent license is “discriminatory” if it does not include within the scope of its coverage, prohibits the exercise of, or is conditioned on the non-exercise of one or more of the rights that are specifically granted under this License. You may not convey a~covered work if you are a~party to an arrangement with a~third party that is in the business of distributing software, under which you make payment to the third party based on the extent of your activity of conveying the work, and under which the third party grants, to any of the parties who would receive the covered work from you, a~discriminatory patent license (a) in connection with copies of the covered work conveyed by you (or copies made from those copies), or (b) primarily for and in connection with specific products or compilations that contain the covered work, unless you entered into that arrangement, or that patent license was granted, prior to 28 March 2007.

Nothing in this License shall be construed as excluding or limiting any implied license or other defenses to infringement that may otherwise be available to you under applicable patent law.

\subsubsection{12. No Surrender of Others' Freedom}

If conditions are imposed on you (whether by court order, agreement or otherwise) that contradict the conditions of this License, they do not excuse you from the conditions of this License. If you cannot convey a~covered work so as to satisfy simultaneously your obligations under this License and any other pertinent obligations, then as a~consequence you may not convey it at all. For example, if you agree to terms that obligate you to collect a~royalty for further conveying from those to whom you convey the Program, the only way you could satisfy both those terms and this License would be to refrain entirely from conveying the Program.

\subsubsection{13. Use with the GNU Affero General Public License}

Notwithstanding any other provision of this License, you have permission to link or combine any covered work with a~work licensed under version 3~of the GNU Affero General Public License into a~single combined work, and to convey the resulting work. The terms of this License will continue to apply to the part which is the covered work, but the special requirements of the GNU Affero General Public License, section 13, concerning interaction through a~network will apply to the combination as such.

\subsubsection{14. Revised Versions of this License}

The Free Software Foundation may publish revised and/or new versions of the GNU General Public License from time to time. Such new versions will be similar in spirit to the present version, but may differ in detail to address new problems or concerns.

Each version is given a~distinguishing version number. If the Program specifies that a~certain numbered version of the GNU General Public License “or any later version” applies to it, you have the option of following the terms and conditions either of that numbered version or of any later version published by the Free Software Foundation. If the Program does not specify a~version number of the GNU General Public License, you may choose any version ever published by the Free Software Foundation.

If the Program specifies that a~proxy can decide which future versions of the GNU General Public License can be used, that proxy's public statement of acceptance of a~version permanently authorizes you to choose that version for the Program.

Later license versions may give you additional or different permissions. However, no additional obligations are imposed on any author or copyright holder as a~result of your choosing to follow a~later version.

\subsubsection{15. Disclaimer of Warranty}

THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM “AS IS” WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A~PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM IS WITH YOU. SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR OR CORRECTION.

\subsubsection{16. Limitation of Liability}

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MODIFIES AND/OR CONVEYS THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A~FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.

\subsubsection{17. Interpretation of Sections 15 and 16}

If the disclaimer of warranty and limitation of liability provided above cannot be given local legal effect according to their terms, reviewing courts shall apply local law that most closely approximates an absolute waiver of all civil liability in connection with the Program, unless a~warranty or assumption of liability accompanies a~copy of the Program in return for a~fee.

\subsection{GNU General Public License, Version 2, June 1991}

Copyright (C) 1989, 1991 Free Software Foundation, Inc. 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA

Everyone is permitted to copy and distribute verbatim copies of this license document, but changing it is not allowed.

\subsubsection{Preamble}

The licenses for most software are designed to take away your freedom to share and change it. By contrast, the GNU General Public License is intended to guarantee your freedom to share and change free software--to make sure the software is free for all its users. This General Public License applies to most of the Free Software Foundation's software and to any other program whose authors commit to using it. (Some other Free Software Foundation software is covered by the GNU Library General Public License instead.) You can apply it to your programs, too.

When we speak of free software, we are referring to freedom, not price. Our General Public Licenses are designed to make sure that you have the freedom to distribute copies of free software (and charge for this service if you wish), that you receive source code or can get it if you want it, that you can change the software or use pieces of it in new free programs; and that you know you can do these things.

To protect your rights, we need to make restrictions that forbid anyone to deny you these rights or to ask you to surrender the rights. These restrictions translate to certain responsibilities for you if you distribute copies of the software, or if you modify it.

For example, if you distribute copies of such a~program, whether gratis or for a~fee, you must give the recipients all the rights that you have. You must make sure that they, too, receive or can get the source code. And you must show them these terms so they know their rights.

We protect your rights with two steps: (1) copyright the software, and (2) offer you this license which gives you legal permission to copy, distribute and/or modify the software.

Also, for each author's protection and ours, we want to make certain that everyone understands that there is no warranty for this free software. If the software is modified by someone else and passed on, we want its recipients to know that what they have is not the original, so that any problems introduced by others will not reflect on the original authors' reputations.

Finally, any free program is threatened constantly by software patents. We wish to avoid the danger that redistributors of a~free program will individually obtain patent licenses, in effect making the program proprietary. To prevent this, we have made it clear that any patent must be licensed for everyone's free use or not licensed at all.

The precise terms and conditions for copying, distribution and modification follow.

\subsubsection{Terms and Conditions for Copying, Distribution and Modification}

0. This License applies to any program or other work which contains a~notice placed by the copyright holder saying it may be distributed under the terms of this General Public License. The “Program“, below, refers to any such program or work, and a~“work based on the Program“ means either the Program or any derivative work under copyright law: that is to say, a~work containing the Program or a~portion of it, either verbatim or with modifications and/or translated into another language. (Hereinafter, translation is included without limitation in the term “modification“.) Each licensee is addressed as “you“.

Activities other than copying, distribution and modification are not covered by this License; they are outside its scope. The act of running the Program is not restricted, and the output from the Program is covered only if its contents constitute a~work based on the Program (independent of having been made by running the Program). Whether that is true depends on what the Program does.

1. You may copy and distribute verbatim copies of the Program's source code as you receive it, in any medium, provided that you conspicuously and appropriately publish on each copy an appropriate copyright notice and disclaimer of warranty; keep intact all the notices that refer to this License and to the absence of any warranty; and give any other recipients of the Program a~copy of this License along with the Program.

You may charge a~fee for the physical act of transferring a~copy, and you may at your option offer warranty protection in exchange for a~fee.

2. You may modify your copy or copies of the Program or any portion of it, thus forming a~work based on the Program, and copy and distribute such modifications or work under the terms of Section 1~above, provided that you also meet all of these conditions:

\begin{enumerate}[label=\Alph*)]
\item You must cause the modified files to carry prominent notices stating that you changed the files and the date of any change.
\item You must cause any work that you distribute or publish, that in whole or in part contains or is derived from the Program or any part thereof, to be licensed as a~whole at no charge to all third parties under the terms of this License.
\item If the modified program normally reads commands interactively when run, you must cause it, when started running for such interactive use in the most ordinary way, to print or display an announcement including an appropriate copyright notice and a~notice that there is no warranty (or else, saying that you provide a~warranty) and that users may redistribute the program under these conditions, and telling the user how to view a~copy of this License. (Exception: if the Program itself is interactive but does not normally print such an announcement, your work based on the Program is not required to print an announcement.)
\end{enumerate}

These requirements apply to the modified work as a~whole. If identifiable sections of that work are not derived from the Program, and can be reasonably considered independent and separate works in themselves, then this License, and its terms, do not apply to those sections when you distribute them as separate works. But when you distribute the same sections as part of a~whole which is a~work based on the Program, the distribution of the whole must be on the terms of this License, whose permissions for other licensees extend to the entire whole, and thus to each and every part regardless of who wrote it.

Thus, it is not the intent of this section to claim rights or contest your rights to work written entirely by you; rather, the intent is to exercise the right to control the distribution of derivative or collective works based on the Program.

In addition, mere aggregation of another work not based on the Program with the Program (or with a~work based on the Program) on a~volume of a~storage or distribution medium does not bring the other work under the scope of this License.

3. You may copy and distribute the Program (or a~work based on it, under Section 2) in object code or executable form under the terms of Sections 1~and 2~above provided that you also do one of the following:

\begin{enumerate}[label=\Alph*)]
\item Accompany it with the complete corresponding machine-readable source code, which must be distributed under the terms of Sections 1~and 2~above on a~medium customarily used for software interchange; or,
\item Accompany it with a~written offer, valid for at least three years, to give any third party, for a~charge no more than your cost of physically performing source distribution, a~complete machine-readable copy of the corresponding source code, to be distributed under the terms of Sections 1~and 2~above on a~medium customarily used for software interchange; or,
\item Accompany it with the information you received as to the offer to distribute corresponding source code. (This alternative is allowed only for noncommercial distribution and only if you received the program in object code or executable form with such an offer, in accord with Subsection b above.)
\end{enumerate}

The source code for a~work means the preferred form of the work for making modifications to it. For an executable work, complete source code means all the source code for all modules it contains, plus any associated interface definition files, plus the scripts used to control compilation and installation of the executable. However, as a~special exception, the source code distributed need not include anything that is normally distributed (in either source or binary form) with the major components (compiler, kernel, and so on) of the operating system on which the executable runs, unless that component itself accompanies the executable.

If distribution of executable or object code is made by offering access to copy from a~designated place, then offering equivalent access to copy the source code from the same place counts as distribution of the source code, even though third parties are not compelled to copy the source along with the object code.

4. You may not copy, modify, sublicense, or distribute the Program except as expressly provided under this License. Any attempt otherwise to copy, modify, sublicense or distribute the Program is void, and will automatically terminate your rights under this License. However, parties who have received copies, or rights, from you under this License will not have their licenses terminated so long as such parties remain in full compliance.

5. You are not required to accept this License, since you have not signed it. However, nothing else grants you permission to modify or distribute the Program or its derivative works. These actions are prohibited by law if you do not accept this License. Therefore, by modifying or distributing the Program (or any work based on the Program), you indicate your acceptance of this License to do so, and all its terms and conditions for copying, distributing or modifying the Program or works based on it.

6. Each time you redistribute the Program (or any work based on the Program), the recipient automatically receives a~license from the original licensor to copy, distribute or modify the Program subject to these terms and conditions. You may not impose any further restrictions on the recipients' exercise of the rights granted herein. You are not responsible for enforcing compliance by third parties to this License.

7. If, as a~consequence of a~court judgment or allegation of patent infringement or for any other reason (not limited to patent issues), conditions are imposed on you (whether by court order, agreement or otherwise) that contradict the conditions of this License, they do not excuse you from the conditions of this License. If you cannot distribute so as to satisfy simultaneously your obligations under this License and any other pertinent obligations, then as a~consequence you may not distribute the Program at all. For example, if a~patent license would not permit royalty-free redistribution of the Program by all those who receive copies directly or indirectly through you, then the only way you could satisfy both it and this License would be to refrain entirely from distribution of the Program.

If any portion of this section is held invalid or unenforceable under any particular circumstance, the balance of the section is intended to apply and the section as a~whole is intended to apply in other circumstances.

It is not the purpose of this section to induce you to infringe any patents or other property right claims or to contest validity of any such claims; this section has the sole purpose of protecting the integrity of the free software distribution system, which is implemented by public license practices. Many people have made generous contributions to the wide range of software distributed through that system in reliance on consistent application of that system; it is up to the author/donor to decide if he or she is willing to distribute software through any other system and a~licensee cannot impose that choice.

This section is intended to make thoroughly clear what is believed to be a~consequence of the rest of this License.

8. If the distribution and/or use of the Program is restricted in certain countries either by patents or by copyrighted interfaces, the original copyright holder who places the Program under this License may add an explicit geographical distribution limitation excluding those countries, so that distribution is permitted only in or among countries not thus excluded. In such case, this License incorporates the limitation as if written in the body of this License.

9. The Free Software Foundation may publish revised and/or new versions of the General Public License from time to time. Such new versions will be similar in spirit to the present version, but may differ in detail to address new problems or concerns.

Each version is given a~distinguishing version number. If the Program specifies a~version number of this License which applies to it and “any later version“, you have the option of following the terms and conditions either of that version or of any later version published by the Free Software Foundation. If the Program does not specify a~version number of this License, you may choose any version ever published by the Free Software Foundation.

10. If you wish to incorporate parts of the Program into other free programs whose distribution conditions are different, write to the author to ask for permission. For software which is copyrighted by the Free Software Foundation, write to the Free Software Foundation; we sometimes make exceptions for this. Our decision will be guided by the two goals of preserving the free status of all derivatives of our free software and of promoting the sharing and reuse of software generally.

\subsubsection{No Warranty}

11. BECAUSE THE PROGRAM IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM “AS IS“ WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A~PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM IS WITH YOU. SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR OR CORRECTION.

12. IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR REDISTRIBUTE THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A~FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.

\subsection{MIT License}

Permission is hereby granted, free of charge, to any person obtaining a~copy of this software and associated documentation files (the “Software“), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS“, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A~PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

\endgroup

% Reference list
\bibliography{sondovac_manual}
\addcontentsline{toc}{section}{References}

\vfill
\hrule
\vfill
\begin{tiny}
  Created in typographical system \hologo{XeLaTeX}, \url{https://en.wikipedia.org/wiki/XeTeX}, references with \hologo{BibTeX}, \url{http://www.bibtex.org/} on openSUSE GNU/Linux, \url{https://www.opensuse.org/}, \today.
\end{tiny}

\end{document}
